---
title: "Linear regression analysis for chromosome 21"
output: html_notebook
---

## Setup
```{r}
myPath <-  #ADD PATH TO GITHUB FOLDER
suppressPackageStartupMessages(library("limma"))
suppressPackageStartupMessages(library("rlist"))
suppressPackageStartupMessages(library("broom"))
library(ggplot2)

geneAnno <- read.delim(file=paste0(myPath,"Annotations/geneAnno_proteinCoding_all_20210503.txt"), stringsAsFactors = FALSE)
colnames(geneAnno)[9] <- "Gene"
geneAnno_x <- geneAnno[geneAnno$chr == "chrX",]
geneAnno_y <- geneAnno[geneAnno$chr == "chrY",]
X_genes_all <-as.character(geneAnno_x$Gene)
Y_genes_all <-as.character(geneAnno_y$Gene)
sexChrom_genes <- c(X_genes_all,Y_genes_all)
PAR1_genes <- geneAnno_x[geneAnno_x$start < 2691188,]$Gene
PAR2_genes <- c("SPRY3","VAMP7","IL9R","WASH6P")
PAR_genes_all <- c(PAR1_genes,PAR2_genes)
NPX_genes <- X_genes_all[! X_genes_all %in% PAR_genes_all]
NPY_genes <- Y_genes_all[! Y_genes_all %in% PAR_genes_all]
autosome_genes <- geneAnno[! geneAnno$chr %in% c("chrX","chrY"),"Gene"]

#bring in chr21 annotation
chr21_full_table <- read.delim(file=paste0(myPath,"Linear_regressions/Chr21/chr21_gene_annotation.txt"),stringsAsFactors = FALSE)
chr21_genes <- as.character(chr21_full_table$gene_name.84)

#read in the metadata table
metadata <-read.delim(paste0(myPath,"Linear_regressions/metadata.txt"), stringsAsFactors = FALSE)
rownames(metadata) <- metadata$sample

#restrict to cell type and karyotype
metadata_tri21 <- metadata[metadata$karyotype %in% c("XX","XY","XX_t21","XY_t21") & metadata$cell_type == "LCL" & metadata$Technical_replicate == "N",c("sample","Chr21_count","sex","batch_libprep","karyotype")]
metadata_tri21$Chr21_count <- metadata_tri21$Chr21_count - 2


chr21Col <- "#DA70D6"
chr21Col_light <- "#DA70D620"

ggGenePlot_regLine_Chr21_count <- function(Gene,all_data, normBy, myWidth=1.75, myHeight=1.75){
   all_data$gene_expression <- all_data$gene_expression / normBy
    all_data$Chr21_count <- all_data$Chr21_count + 2
  
  pdf(file=paste0(Gene,"_linearPlot_chr21.pdf"), width=myWidth, height=myHeight)
  p <- ggplot(data=all_data, mapping = aes(x=Chr21_count, y=gene_expression)) +
    geom_jitter(color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(method = "lm", color=chr21Col, size=0.25, fill=chr21Col_light, fullrange=TRUE) +
    labs( 
      x= "Number of Chr 21",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(-0.01,max(all_data$gene_expression)+ 0.5) + xlim(1.7,3.3)
  print(p)
  dev.off()
}
ggGenePlot_regLine_Chr21_count_max <- function(Gene,all_data, normBy, myWidth=1.75, myHeight=1.75, myYmax = 10){
   all_data$gene_expression <- all_data$gene_expression / normBy
    all_data$Chr21_count <- all_data$Chr21_count + 2
  
  pdf(file=paste0(Gene,"_linearPlot_chr21.pdf"), width=myWidth, height=myHeight)
  p <- ggplot(data=all_data, mapping = aes(x=Chr21_count, y=gene_expression)) +
    geom_jitter(color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(method = "lm", color=chr21Col, size=0.25, fill=chr21Col_light, fullrange=TRUE) +
    labs( 
      x= "Number of Chr 21",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(-0.01,myYmax) + xlim(1.7,3.3)
  print(p)
  dev.off()
}
```

#Processing raw RNA-seq data
```{r, eval=FALSE}
### The following code is used to run this analysis from the raw data. Skip below to begin with provided processed data file. ###
## You must obtain access to the raw data through dbGaP (accession # phs002481). Download the fastq files, and run kallisto using the following command:
# kallisto quant -i KALLISTO_INDEX_FILE -t 16 --bias --plaintext -o  /kallisto_OUTPUT_FOLDER/sampleName/ file1fastq.gz file2fastq.gz 

#Bring in list of Kallisto output files
myFileTable <- read.table(file =  "KALLISTO_OUTPUT_FILE_PATH", row.names = 1, stringsAsFactors = FALSE)
myFiles <- myFileTable$V2
names(myFiles) <- rownames(myFileTable)
#subset file list for the samples I will use in this analysis
myFiles_tri21 <- myFiles[metadata_tri21$sample]

#run tximport to bring in mapped samples
suppressPackageStartupMessages(library("tximport"))

annofile <-read.delim(file=paste0(myPath,"Annotations/annotation_with_ERCC.txt"),sep = " ")
tx2gene <-data.frame("TXNAME" = annofile$transcript_id,"GENEID" = annofile$gene_name)

txi_tri21 <-tximport(myFiles_tri21, type = "kallisto", tx2gene = tx2gene)
 

# #Adjust sampleTable to match list of samples in txi
sample_table_tri21 <- metadata_tri21[colnames(txi_tri21$counts), ]


# #use scientific notation
options(scipen = 0, digits = 5)

#make sure chr21_count, and batch_libprep is a factor
class(sample_table_tri21$chr21_count)
sample_table_tri21$batch_libprep <- as.factor(sample_table_tri21$batch_libprep)
class(sample_table_tri21$batch_libprep)

# Get list of expressed genes in the cell type of interest
XX_sample_names <- metadata_tri21[metadata_tri21$karyotype == "XX", "sample"]
XY_sample_names <- metadata_tri21[metadata_tri21$karyotype == "XY", "sample"]
tpm1_genes_xx_xy_median <- function(tximport_file, XX_samples, XY_samples) {
  kallisto_output_TPM_XY <- tximport_file$abundance[,c(XY_samples)]
  kallisto_output_TPM_XX <- tximport_file$abundance[,c(XX_samples)]
  median_TPM_XY <- apply(kallisto_output_TPM_XY,1,median)
  median_TPM_XX <- apply(kallisto_output_TPM_XX,1,median)
  #Put all of the medians together in a dataframe of expression per karyotype - numeric
  expression_perKaryotype <- data.frame(median_TPM_XY, median_TPM_XX)
  #Isolate the rows of the expression_TPM dataset that have at least one karyotype passing the threshold
  expressedGenes <- expression_perKaryotype[expression_perKaryotype$median_TPM_XY >= 1 | expression_perKaryotype$median_TPM_XX >= 1,]
  #Restrict by gene annotation file
  expressedGenes <- expressedGenes[rownames(expressedGenes) %in% geneAnno$Gene,]
  #write.table(expressedGenes, file=paste0("expressedGenes_TPM_",tximport_file,".txt"),quote=FALSE,sep="\t")
  #See how many genes are expressed in at least one karyotype
  num_expressedGenes <- dim(expressedGenes)[1]
  print(paste("Number of 'expressed' genes with median TPM >= 1 in either XX or XY samples: ", as.character(num_expressedGenes), sep=""))

  #Find expressed sex chromosome genes
  expSexChromGenes <- expressedGenes[rownames(expressedGenes) %in% sexChrom_genes,]
  print(paste("Number of 'expressed' sex chromosome genes: ", as.character(dim(expSexChromGenes)[1]), sep=""))
  #write.table(expSexChromGenes, file="expressedSexChromGenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed Y chromosome genes
  expYGenes <- expressedGenes[rownames(expressedGenes) %in% Y_genes_all,]
  print(paste("Number of 'expressed' Y chromosome genes: ", as.character(dim(expYGenes)[1]), sep=""))
  #write.table(expYGenes, file="expressedYgenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed X chromosome genes
  expXGenes <- expressedGenes[rownames(expressedGenes) %in% X_genes_all,]
  print(paste("Number of 'expressed' X chromosome genes: ", as.character(dim(expXGenes)[1]), sep=""))
  #write.table(expXGenes, file="expressedXgenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed PAR genes
  expPARGenes <- expressedGenes[rownames(expressedGenes) %in% PAR_genes_all,]
  print(paste("Number of 'expressed' PAR genes (small set): ", as.character(dim(expPARGenes)[1]), sep=""))
  #write.table(expPARGenes, file="expressedPARgenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed autosome genes
  expAutoGenes <- expressedGenes[rownames(expressedGenes) %in% autosome_genes,]
  print(paste("Number of 'expressed' autosome genes: ", as.character(dim(expAutoGenes)[1]), sep=""))
  #write.table(expAutoGenes, file="expressedAutosomeGenes_TPM.txt",quote=FALSE,sep="\t")

  myList <- list("expressedGenes" = rownames(expressedGenes), "expSexChromGenes" = rownames(expSexChromGenes),
                 "expYGenes" = rownames(expYGenes), "expXGenes" = rownames(expXGenes), "expPARGenes" = rownames(expPARGenes),
                 "expAutoGenes" = rownames(expAutoGenes))

  return(myList)
}

expressed <- tpm1_genes_xx_xy_median(tximport_file = txi_tri21,XX_samples = XX_sample_names, XY_samples = XY_sample_names )

chr21_exp <- intersect(expressed$expressedGenes, chr21_genes)

#run deseq to get the normalized counts matrix across all samples.
suppressPackageStartupMessages(library("DESeq2"))
dds_tri21 <- DESeqDataSetFromTximport(txi_tri21, colData = sample_table_tri21, design = ~ batch_libprep + Chr21_count + sex)
dds_tri21 <- estimateSizeFactors(object = dds_tri21)

normCounts_tri21 <- counts(dds_tri21, normalized=TRUE)
normCounts_exp21genes <- normCounts_tri21[rownames(normCounts_tri21) %in% chr21_exp,]
write.table(normCounts_exp21genes, file="lcl_normCounts_chr21_genes.txt", quote=FALSE, sep="\t", row.names=TRUE, col.names=TRUE)
```

## Linear modeling of chr21 genes
```{r}
## START HERE WITH PROCESSED NORMALIZED COUNTS DATA
normCounts_exp21genes <- read.delim(file=paste0(myPath,"Linear_regressions/Chr21/lcl_normCounts_chr21_genes.txt"), stringsAsFactors = FALSE, check.names = FALSE)

#make sure metadata is in same order as the normalized counts data!
my_metadata_tri21 <- metadata_tri21[colnames(normCounts_exp21genes),c("Chr21_count", "sex", "batch_libprep")]

special <- c("CCT8")
#For top significant genes
# special <- c("HLCS","CRYZL1","RWDD2B","PFKL","TRAPPC10","TIAM1","MCM3AP","PCNT","USP16","C21orf2","TMEM50B","URB1","DIP2A","IFNAR1","SYNJ1")

myResults_chr21 <- data.frame(intercept=numeric(),intErr = numeric(),chr21beta=numeric(), chr21pval = numeric(), chr21Err = numeric(),sexbeta=numeric(), sexpval=numeric(),sexErr=numeric(),adj_rsq=numeric(),pval=numeric())

for(gene in rownames(normCounts_exp21genes)){
    #pull out data
    myGeneData <- t(normCounts_exp21genes[gene,])
    all_Data <- cbind(my_metadata_tri21, myGeneData)
    colnames(all_Data) <- c("Chr21_count", "sex", "batch_libprep", "gene_expression")
    #do a linear regression
    myFormula <- formula(gene_expression ~ Chr21_count + sex + batch_libprep)
    myFormula_int <- formula(gene_expression ~ Chr21_count + sex)
    mylm <- lm(myFormula, data=all_Data)
    mylm_int <- lm(myFormula_int, data=all_Data)
    #pull out beta, p-value and add to table
    intercept <- (summary(mylm_int)$coefficients["(Intercept)",1])/2 #divide by two to get the value for one chr21
    interr <- summary(mylm_int)$coefficients["(Intercept)",2]
    chr21beta <- summary(mylm)$coefficients["Chr21_count",1]
    chr21pval <- summary(mylm)$coefficients["Chr21_count",4]
    chr21err <- summary(mylm)$coefficients["Chr21_count",2]
    sexbeta <- summary(mylm)$coefficients["sexM",1]
    sexpval <- summary(mylm)$coefficients["sexM",4]
    sexerr <- summary(mylm)$coefficients["sexM",2]
    adj_rsq <- glance(mylm)$adj.r.squared
    pval <- glance(mylm)$p.value
 
    myResults_chr21[gene,] <- c(intercept,interr,chr21beta, chr21pval, chr21err,sexbeta, sexpval, sexerr,adj_rsq,pval)
    
    #if gene in special list
  if(gene %in% special){
      ggGenePlot_regLine_Chr21_count(Gene = gene, all_data = all_Data,normBy = 1000)
  }
    if(gene %in% "PFKL"){
      ggGenePlot_regLine_Chr21_count_max(Gene = gene, all_data = all_Data,normBy = 1000, myYmax = 30)
    }
    if(gene %in% "TRAPPC10"){
      ggGenePlot_regLine_Chr21_count_max(Gene = gene, all_data = all_Data,normBy = 1000, myYmax = 9)
    }
    if(gene %in% "URB1"){
      ggGenePlot_regLine_Chr21_count_max(Gene = gene, all_data = all_Data,normBy = 1000, myYmax = 7)
  }
}

#correct for multiple hypothesis testing and set signficance threshold to padj < 0.05
options(scipen = 0) 
bh_adj <- p.adjust(myResults_chr21$chr21pval, method="BH")
bh_adj_sex <- p.adjust(myResults_chr21$sexpval, method="BH")
bh_adj_reg <- p.adjust(myResults_chr21$pval, method="BH")
deltaChr21 <- myResults_chr21$chr21beta/myResults_chr21$intercept

deltaChr21_err <- sqrt(((myResults_chr21$chr21beta ^ 2)/(myResults_chr21$intercept ^ 2)) * 
  (((myResults_chr21$chr21Err ^ 2)/(myResults_chr21$chr21beta ^ 2)) + ((myResults_chr21$intErr ^ 2)/(myResults_chr21$intercept ^ 2))))

deltaEchr21_95_CI_lo <- deltaChr21 - (1.96 * deltaChr21_err)
deltaEchr21_95_CI_hi <- deltaChr21 + (1.96 * deltaChr21_err)

deltaSex <- myResults_chr21$sexbeta/myResults_chr21$intercept
deltaSex_err <- ((myResults_chr21$sexbeta ^ 2)/(myResults_chr21$intercept ^ 2)) * 
  (((myResults_chr21$sexErr ^ 2)/(myResults_chr21$sexbeta ^ 2)) + ((myResults_chr21$intErr ^ 2)/(myResults_chr21$intercept ^ 2)))

deltaEsex_95_CI_lo <- deltaSex - (1.96 * deltaSex_err)
deltaEsex_95_CI_hi <- deltaSex + (1.96 * deltaSex_err)

myResults_chr21_delta21 <- data.frame(row.names= rownames(myResults_chr21),
                                      "gene" = rownames(myResults_chr21),
                          "intercept" = myResults_chr21$intercept,
                          "intercept_err" = myResults_chr21$intErr,
                          "chr21_coeff" = myResults_chr21$chr21beta,
                          "chr21_coeff_err" = myResults_chr21$chr21Err,
                          "chr21_pval" = myResults_chr21$chr21pval,
                          "chr21_adj_pval" = bh_adj,
                          "deltaEchr21" = deltaChr21,
                          "deltaEchr21_err" = deltaChr21_err,
                          "deltaEchr21_95_CI_lo" = deltaEchr21_95_CI_lo,
                          "deltaEchr21_95_CI_hi" = deltaEchr21_95_CI_hi,
                          "sex_coeff" = myResults_chr21$sexbeta,
                          "sex_coeff_err" = myResults_chr21$sexErr,
                          "sex_pval" = myResults_chr21$sexpval,
                          "sex_adj_pval" = bh_adj_sex,
                          "deltaEsex" = deltaSex,
                          "deltaEsex_err" = deltaSex_err,
                          "deltaEsex_95_CI_lo" = deltaEsex_95_CI_lo,
                          "deltaEsex_95_CI_hi" = deltaEsex_95_CI_hi,
                          "adj.r.sq" = myResults_chr21$adj_rsq,
                          "reg_pval" = myResults_chr21$pval,
                          "reg_adj_pval" = bh_adj_reg
                          )
write.table(x = myResults_chr21_delta21, file = "LCL_regression_results_chr21.txt", quote = FALSE, sep = "\t", row.names=FALSE, col.names=TRUE)
```