---
title: "Sleuth linear regression"
author: Hannah Harris
---

#load libraries
```{r}
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(tidyr))
suppressPackageStartupMessages(library("limma"))
suppressPackageStartupMessages(library("broom"))
suppressPackageStartupMessages(library(sleuth))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(ggpubr))
```

#process metadata
```{r}
process_metadata <- function(metadata, cell){
  #takes in a cell type 
  #processes metadata 
  rownames(metadata) <- metadata$sample
  #subtract 1 from X count to model number of additional X chromosomes
  metadata$x_count <- metadata$x_count - 1
metadata <- metadata[metadata$karyotype %in% c("X","XX","XXX","XXXX","XXXXY","XXXY","XXY","XXYY","XY","XYY","XYYYY") & metadata$cell_type == cell & metadata$Technical_replicate == "N",]
  return(metadata)
}
```

#sleuth prep function
```{r}

sleuth_normalize <- function(metadata, tx2gene){
  #runs sleuth prep function and returns sleuth object and normalized count table
  
  so <- sleuth_prep(metadata, target_mapping = tx2gene, extra_bootstrap_summary = TRUE, aggregation_column = 'ens_gene')
  #extract normalized counts 
  normalized <- kallisto_table(so, normalized = TRUE) #filtered == TRUE
  cut_normalized <- normalized[1:3]
  pivotted_data <- pivot_wider(cut_normalized,names_from = sample, values_from = est_counts)
  rownames <- pivotted_data$target_id
  counts_df <- pivotted_data[,-1]
  row.names(counts_df) <- rownames
  
  return(list(so = so, counts = counts_df))
}

```

#run sleuth
```{r}
run_sleuth <- function(so){
  #runs sleuth models
  
  so <- sleuth_fit(so, ~y_count + batch_libprep, 'reduced')
  so <- sleuth_fit(so, ~x_count + y_count + batch_libprep, 'full')
  so <- sleuth_lrt(so, 'reduced', 'full')
  
  sleuth_table_tx <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE, pval_aggregate = FALSE)
  
  return(list(so = so, table_tx = sleuth_table_tx)) 
}
```

#linear regression function
```{r}
linear_regression <-  function(norm_counts_df, tx2gene, metadata, geneAnno, XIST_expr = TRUE) {
    #runs linear regresssion and return results table 
  
    #get gene lists
    colnames(geneAnno)[9] <- "Gene"
    geneAnno_x <- geneAnno[geneAnno$chr == "chrX",]
    geneAnno_y <- geneAnno[geneAnno$chr == "chrY",]
    X_genes_all <-as.character(geneAnno_x$Gene)
    PAR1_genes <- geneAnno_x[geneAnno_x$start < 2691190,]$Gene
    PAR2_genes <- c("SPRY3","VAMP7","IL9R","WASH6P")
    PAR_genes_all <- c(PAR1_genes,PAR2_genes)
    NPX_genes <- X_genes_all[! X_genes_all %in% PAR_genes_all]


  metadata_modelling <- metadata[,c("x_count","y_count","batch_libprep")]

myResults <- data.frame(intercept=numeric(),intErr = numeric(),xbeta=numeric(), xpval = numeric(), xErr = numeric(),ybeta=numeric(), ypval=numeric(),yErr=numeric(), adj_rsq=numeric(), pval=numeric())
region_res <- character()
gene_names <- character()

for(transcript in rownames(norm_counts_df)){
    options(scipen = 0, digits = 5) 
    #pull out data
    myGeneData <- as.data.frame(t(norm_counts_df[transcript,]))


  
    all_Data <- merge(metadata_modelling, myGeneData, by = "row.names")
    row.names(all_Data) <- all_Data$Row.names
    all_Data <- all_Data[,2:5]
    colnames(all_Data) <- c("x_count", "y_count", "batch_libprep", "gene_expression")
    #do a linear regression
    myFormula <- formula(gene_expression ~ x_count + y_count + batch_libprep)
    myFormula_int <- formula(gene_expression ~ x_count + y_count )
    mylm <- lm(myFormula, data=all_Data)
    mylm_int <- lm(myFormula_int, data=all_Data)
    #pull out intercept, betas, p-values and add to table
    intercept <- summary(mylm_int)$coefficients["(Intercept)",1]
    interr <- summary(mylm_int)$coefficients["(Intercept)",2]
    xbeta <- summary(mylm)$coefficients["x_count",1]
    xpval <- summary(mylm)$coefficients["x_count",4]
    xerr <- summary(mylm)$coefficients["x_count",2]
    ybeta <- summary(mylm)$coefficients["y_count",1]
    ypval <- summary(mylm)$coefficients["y_count",4]
    yerr <- summary(mylm)$coefficients["y_count",2]
    adj_rsq <- glance(mylm)$adj.r.squared
    pval <- glance(mylm)$p.value
  
    #label genes with regions  
    
    crop_tx2gene <- tx2gene %>%  filter(target_id == transcript)
    gene_ext <- as.character(crop_tx2gene$ext_gene) 
  
      if(gene_ext %in% NPX_genes){
        region <- "NPX"
      }
      else if(gene_ext %in% PAR1_genes){
        region <- "PAR1"
  }
      else if(gene_ext %in% PAR2_genes){
        region <- "PAR2"
  }
      else{
        region <-  "NPY"
  }

  #build results into table  
  myResults[transcript,] <- c(intercept,interr,xbeta, xpval, xerr,ybeta, ypval, yerr, adj_rsq,pval)
  region_res <- c(region_res, region) 
  gene_names <-  c(gene_names, gene_ext)
}

myResults_par_npx <- cbind(myResults,region_res,gene_names)
#myResults_par_npx <- cbind(myResults,region_res)

#ADRIANNA CODE BELOW:
#correct for multiple hypothesis testing and set signficance threshold to padj < 0.05
bh_adj <- p.adjust(myResults_par_npx$xpval, method="BH")
bh_adj_y <- p.adjust(myResults_par_npx$ypval, method="BH")
bh_adj_reg <- p.adjust(myResults_par_npx$pval, method="BH")
deltaEx <- myResults_par_npx$xbeta/myResults_par_npx$intercept
deltaEx_err <- sqrt(
  ((myResults_par_npx$xbeta ^ 2)/(myResults_par_npx$intercept ^ 2)) * 
  (
    ((myResults_par_npx$xErr ^ 2)/(myResults_par_npx$xbeta ^ 2)) + 
      ((myResults_par_npx$intErr ^ 2)/(myResults_par_npx$intercept ^ 2))
    )
  )
deltaEy <- myResults_par_npx$ybeta/myResults_par_npx$intercept
deltaEy_err <- sqrt(
  ((myResults_par_npx$ybeta ^ 2)/(myResults_par_npx$intercept ^ 2)) * 
  (
    ((myResults_par_npx$yErr ^ 2)/(myResults_par_npx$ybeta ^ 2)) + 
      ((myResults_par_npx$intErr ^ 2)/(myResults_par_npx$intercept ^ 2))
    )
  )

#put together final table
myResults_npx_par_deltaEx_y <- data.frame(row.names= rownames(myResults_par_npx),
                          "gene"= myResults_par_npx$gene_names,
                          "intercept" = myResults_par_npx$intercept,
                          "intercept_err" = myResults_par_npx$intErr,
                          "x_coeff" = myResults_par_npx$xbeta,
                          "x_coeff_err" = myResults_par_npx$xErr,
                          "x_pval" = myResults_par_npx$xpval,
                          "x_adj_pval" = bh_adj,
                          "y_coeff" = myResults_par_npx$ybeta,
                          "y_coeff_err" = myResults_par_npx$yErr,
                          "y_pval" = myResults_par_npx$ypval,
                          "y_adj_pval" = bh_adj_y,
                          "adj.r.sq" = myResults_par_npx$adj_rsq,
                          "reg_pval" = myResults_par_npx$pval,
                          "reg_adj_pval" = bh_adj_reg,
                          "deltaEx" = deltaEx,
                          "deltaEx_err" = deltaEx_err,
                          "deltaEy" = deltaEy,
                          "deltaEy_err" = deltaEy_err,
                          "region" = myResults_par_npx$region_res
                          )

#adjust for XIST:
if (XIST_expr) {
  print("IM HERE")
  XIST_int <- myResults_par_npx["XIST","xbeta"] + myResults_par_npx["XIST","intercept"]
  deltaEx_XIST <- myResults_par_npx["XIST","xbeta"]/XIST_int
  deltaEy_XIST <- myResults_par_npx["XIST","ybeta"]/XIST_int

  deltaEx_err_XIST <- sqrt(
  ((myResults_par_npx["XIST","xbeta"] ^ 2)/(XIST_int ^ 2)) * 
  (
    ((myResults_par_npx["XIST","xErr"] ^ 2)/(myResults_par_npx["XIST","xbeta"] ^ 2)) + 
      (((myResults_par_npx["XIST","intErr"] ^ 2) + (myResults_par_npx["XIST","xErr"] ^ 2))/(XIST_int ^ 2))
    )
  )

  deltaEy_err_XIST <- sqrt(
  ((myResults_par_npx["XIST","ybeta"] ^ 2)/(XIST_int ^ 2)) * 
  (
    ((myResults_par_npx["XIST","yErr"] ^ 2)/(myResults_par_npx["XIST","ybeta"] ^ 2)) + 
      (((myResults_par_npx["XIST","intErr"] ^2) + (myResults_par_npx["XIST","yErr"] ^ 2 ))/(XIST_int ^ 2))
    )
  )


  myResults_npx_par_deltaEx_y["XIST",c("deltaEx","deltaEx_err","deltaEy","deltaEy_err")] <- c("deltaEx" = deltaEx_XIST, "deltaEx_err" = deltaEx_err_XIST,"deltaEy" = deltaEy_XIST,"deltaEy_err" = deltaEy_err_XIST)
  
  
}

return(myResults_npx_par_deltaEx_y)
}
```

#scatter plots 
```{r}
scatter_df <- function(regr_fib_X_pval, filename, width = 5, height = 5){
  
regr_fib_X_pval <- regr_fib_X_pval 

genes <- as.character(unique(regr_fib_X_pval$gene))
myResults <- data.frame(gene = character(),transcript1_Ex=numeric(), transcript2_Ex = numeric(), gene_deltaEx = numeric(),  fill = character(), alpha = numeric(), stringsAsFactors = FALSE)
row_num = 1
for (gene_id in genes){
    gene_res <- regr_fib_X_pval %>% filter(gene == gene_id) #%>% filter(x_adj_pval < 0.05)
    gene_deltaEx <- as.character(unique(gene_res$gene_deltaEx)) 
    fill <- as.character(unique(gene_res$fill))
    alpha <- as.character(unique(gene_res$alpha)) 

    if (nrow(gene_res) > 1) {
      max_val <- max(gene_res$deltaEx)
      min_val <- min(gene_res$deltaEx)
      for (deltaEx in gene_res$deltaEx){
          myResults[row_num,]<- c(gene_id, max_val, deltaEx, gene_deltaEx, fill, alpha)
          row_num  = row_num + 1
        
      }
     
    }
    
}

# print(myResults)
myResults$transcript1_Ex <- as.numeric(myResults$transcript1_Ex)
myResults$transcript2_Ex <- as.numeric(myResults$transcript2_Ex)
myResults$gene_deltaEx <- as.numeric(myResults$gene_deltaEx)

pdf(filename, width, height)

p <- ggscatter(myResults, x = "gene_deltaEx", y = "transcript2_Ex", color = "fill", alpha = "alpha", add = "reg.line", 
          #label = 'gene', repel = TRUE,
          palette = c("black", "red", "blue"), xlab = "gene deltaEx", ylab = "transcript deltaEx", font.label = c(8, "plain"),ylim=c(-.25,1.5), xlim = c(-.25,1.5)) #+ #label.select = c("UBA1", "EIF1AX")) + 
  #stat_cor(method = "pearson", label.x = 0, label.y = 1)
print(p + rremove("legend"))

dev.off 
}


```


#Main
```{r fig.width=8}
myPath <-  #ADD PATH TO GITHUB FOLDER

#load files
#genes with greater than one transcript
transcript_to_TSS <- read.csv(file = paste0(myPath,'Linear_regressions_transcripts/TSS_annotations_greaterthan1.csv')) 

metadata <- read.delim(file = paste0(myPath,"Linear_regressions/metadata.txt"), check.names = FALSE)

#Load gene-level data
LCL_regression_X_genes <- read.delim(file=paste0(myPath,"/Linear_regressions/LCLs/lcl_regression_results_npx_par.txt"), stringsAsFactors = FALSE)
LCL_regression_X_genes$gene_deltaEx <- LCL_regression_X_genes$deltaEx 
LCL_regression_X_genes <- LCL_regression_X_genes[,c("gene","gene_deltaEx")]
fib_regression_X_genes <- read.delim(file=paste0(myPath,"/Linear_regressions/Fibroblasts/fib_regression_results_npx_par.txt"), stringsAsFactors = FALSE)
fib_regression_X_genes$gene_deltaEx <- fib_regression_X_genes$deltaEx 
fib_regression_X_genes <- fib_regression_X_genes[ ,c("gene","gene_deltaEx")]

Xgenes_LCL <- LCL_regression_X_genes$gene
Xgenes_fib <- fib_regression_X_genes$gene

greatthan1tx <- as.character(unique(transcript_to_TSS$gene_name))
length(intersect(greatthan1tx, Xgenes_LCL))  
length(intersect(greatthan1tx, Xgenes_fib)) 

#Set up sex chromosome gene lists, using gencode v84 as "Gene" list
geneAnno <- read.delim(paste0(myPath,"Annotations/geneAnno_proteinCoding_all_20210503.txt"), stringsAsFactors = FALSE) 

#extract transcripts from gtf 
gtf <- read.table(paste0(myPath,"Annotations/annotation_with_ERCC.txt"), header = TRUE)
tx2gene <- data.frame("target_id"=gtf$transcript_id, "ens_gene"= gtf$gene_id, "ext_gene" = gtf$gene_name)
tx2gene$target_id <- as.character(tx2gene$target_id)

#make subsetted metadata files 
metadata_lcl <- process_metadata(metadata, 'LCL')
metadata_fib <- process_metadata(metadata, 'Fib')


### The following code is used to run this analysis from the raw data. Skip to below to begin with provided processed data file. ###
## You must obtain access to the raw data through dbGaP (accession # phs002481). Download the fastq files, and run kallisto with 100 bootstraps using the following command:
# kallisto quant -i /Annotations/ERCC_gencode_merge.ksto -t 16 --bias --plaintext -o  /kallisto_OUTPUT_FOLDER/sampleName/ file1fastq.gz file2fastq.gz -b 100

### Add paths to kallisto output files to the metadata table:
# metadata$path <- paste("kallisto_OUTPUT_FOLDER/", metadata$sample, "/abundance.h5", sep = "")

#run sleuth 
# sleuth_fib <- sleuth_normalize(metadata_fib, tx2gene)
# sleuth_fib_so <- sleuth_fib$so
# sleuth_fib_list <- run_sleuth(sleuth_fib_so)
# sleuth_fib_so <- sleuth_fib_list$so
# gene_sleuth_fib <- sleuth_fib_list$table_gene
# tx_sleuth_fib <- sleuth_fib_list$table_tx
# 
# sleuth_save(sleuth_fib_so, "sleuth_fib.rda" )
# write.table(x = sleuth_fib$counts, file = "sleuth_fib_normcounts.txt", quote = FALSE, sep = "\t")
# write.table(x = gene_sleuth_fib, file = "gene_sleuth_fib_res.txt", quote = FALSE, sep = "\t")
# write.table(x = tx_sleuth_fib, file = "tx_sleuth_fib_res.txt", quote = FALSE, sep = "\t")
# 
# 
# 
# sleuth_lcl <- sleuth_normalize(metadata_lcl, tx2gene)
# sleuth_lcl_so <- sleuth_lcl$so
# sleuth_lcl_list <- run_sleuth(sleuth_lcl_so)
# sleuth_lcl_so <- sleuth_lcl_list$so
# gene_sleuth_lcl <- sleuth_lcl_list$table_gene
# tx_sleuth_lcl <- sleuth_lcl_list$table_tx
# 
# sleuth_save(sleuth_lcl_so, "sleuth_lcl.rda" )
# write.table(x = sleuth_lcl$counts, file = "sleuth_lcl_normcounts.txt", quote = FALSE, sep = "\t")
# write.table(x = gene_sleuth_lcl, file = "gene_sleuth_lcl_res.txt", quote = FALSE, sep = "\t")
# write.table(x = tx_sleuth_lcl, file = "tx_sleuth_lcl_res.txt", quote = FALSE, sep = "\t")

#run the wald test on the sleuth objects
#sleuth_lcl <- sleuth_load("sleuth_lcl.rda")
#sleuth_fib <- sleuth_load("sleuth_fib.rda")
#sleuth_lcl <- sleuth_wt(sleuth_lcl, 'x_count', which_model = "full")
#sleuth_fib <- sleuth_wt(sleuth_fib, 'x_count', which_model = "full")
#lcl_tx_waldtest <- sleuth_results(sleuth_lcl, 'x_count', test_type = "wt", which_model = "full", pval_aggregate = FALSE)
#fib_tx_waldtest <- sleuth_results(sleuth_fib, 'x_count', test_type = "wt", which_model = "full", pval_aggregate = FALSE)

# write.table(lcl_tx_waldtest, file = "lcl_tx_waldtest.txt", quote = FALSE, sep = "\t")
# write.table(fib_tx_waldtest, file = "fib_tx_waldtest.txt", quote = FALSE, sep = "\t")


#load sleuth normalized count files for transcript isoforms of expressed X chromosome genes with median counts >100
sleuth_lcl_normcounts_X <- read.delim(file = paste0(myPath,"Linear_regressions_transcripts/sleuth_lcl_normcounts_expXisoforms.txt"), check.names = FALSE) 
sleuth_fib_normcounts_X <- read.delim(file = paste0(myPath,"Linear_regressions_transcripts/sleuth_fib_normcounts_expXisoforms.txt"), check.names = FALSE)

#get tables of all expressed transcripts 
lcl_subsetted_transcripts_list <- rownames(sleuth_lcl_normcounts_X)
fib_subsetted_transcripts_list <- rownames(sleuth_fib_normcounts_X)

#do linear regression on all expressed transcripts
regr_lcl_X <- linear_regression(sleuth_lcl_normcounts_X, tx2gene, metadata_lcl, geneAnno, XIST_expr = TRUE)
regr_lcl_X$gene <- as.character(regr_lcl_X$gene)
# write.table(x = regr_lcl_X, file = "regression_lcl_X_isoforms.txt", quote = FALSE, sep = "\t")
regr_fib_X <- linear_regression(sleuth_fib_normcounts_X, tx2gene, metadata_fib, geneAnno, XIST_expr = TRUE)
regr_fib_X$gene <- as.character(regr_fib_X$gene)
# write.table(x = regr_fib_X, file = "regression_fib_X_isoforms.txt", quote = FALSE, sep = "\t")

#number of isoforms per gene
lcl_isoforms_per_gene <- data.frame(table(regr_lcl_X$gene))
fib_isoforms_per_gene <- data.frame(table(regr_fib_X$gene))

#Total genes with more than 1 expressed isoform:
lcl_multi_iso <- as.character(lcl_isoforms_per_gene[lcl_isoforms_per_gene$Freq >1,"Var1"])
length(lcl_multi_iso) #169
fib_multi_iso <- as.character(fib_isoforms_per_gene[fib_isoforms_per_gene$Freq >1,"Var1"])
length(fib_multi_iso) #178

#Subset regression
regr_lcl_X_multi_iso <- regr_lcl_X[regr_lcl_X$gene %in% lcl_multi_iso,]
regr_fib_X_multi_iso <- regr_fib_X[regr_fib_X$gene %in% fib_multi_iso,]

#Significant X-response
regr_lcl_X_multi_iso_sig <- regr_lcl_X_multi_iso[regr_lcl_X_multi_iso$x_adj_pval < 0.05,]
regr_fib_X_multi_iso_sig <- regr_fib_X_multi_iso[regr_fib_X_multi_iso$x_adj_pval < 0.05,]

#Not significant X-response
regr_lcl_X_multi_iso_NS <- regr_lcl_X_multi_iso[regr_lcl_X_multi_iso$x_adj_pval >= 0.05,]
regr_fib_X_multi_iso_NS <- regr_fib_X_multi_iso[regr_fib_X_multi_iso$x_adj_pval >= 0.05,]


#All isoforms deltaEx not sig:
#LCL 
LCL_not_sig_genes <- setdiff(regr_lcl_X_multi_iso_NS$gene, regr_lcl_X_multi_iso_sig$gene)
length(LCL_not_sig_genes)

#Fib
Fib_not_sig_genes <- setdiff(regr_fib_X_multi_iso_NS$gene, regr_fib_X_multi_iso_sig$gene)
length(Fib_not_sig_genes)

#All isoforms deltaEx significant:
#LCL
LCL_sig_genes <- setdiff(regr_lcl_X_multi_iso_sig$gene, regr_lcl_X_multi_iso_NS$gene)
length(LCL_sig_genes)

#Fib
Fib_sig_genes <- setdiff(regr_fib_X_multi_iso_sig$gene, regr_fib_X_multi_iso_NS$gene)
length(Fib_sig_genes)


#Isoforms discordant:
#LCL
LCL_discordant_genes <- intersect(regr_lcl_X_multi_iso_sig$gene, regr_lcl_X_multi_iso_NS$gene)
length(LCL_discordant_genes)

#Fib
Fib_discordant_genes <- intersect(regr_fib_X_multi_iso_sig$gene, regr_fib_X_multi_iso_NS$gene)
length(Fib_discordant_genes)
```

#generating scatter plot of deltaEx values for transcripts versus genes
```{r}
#filter on deltaEx to remove outliers (deltaEx > 1)
regr_lcl_X <- regr_lcl_X %>% filter(deltaEx < 2)
regr_fib_X <- regr_fib_X %>% filter(deltaEx < 2)

joined_regr_lcl_X <-left_join(regr_lcl_X, LCL_regression_X_genes, by = "gene")
joined_regr_fib_X <-left_join(regr_fib_X, fib_regression_X_genes, by = "gene")

joined_regr_lcl_X$fill <- ifelse(joined_regr_lcl_X$gene == "CA5B", "CA5B", ifelse(joined_regr_lcl_X$gene == "UBA1", "UBA1", "black"))
joined_regr_fib_X$fill <- ifelse(joined_regr_fib_X$gene == "CA5B", "CA5B", ifelse(joined_regr_fib_X$gene == "UBA1", "UBA1", "black"))

#alpha_values: 
joined_regr_lcl_X$alpha <- ifelse(joined_regr_lcl_X$gene == "CA5B", 1, ifelse(joined_regr_lcl_X$gene == "UBA1", 1, 0.7))
joined_regr_fib_X$alpha <- ifelse(joined_regr_fib_X$gene == "CA5B", 1, ifelse(joined_regr_fib_X$gene == "UBA1", 1, 0.7))


#generate scatter
scatter_df(joined_regr_fib_X, "fib_scatter.pdf")
scatter_df(joined_regr_lcl_X, "lcl_scatter.pdf")
```

#Graph per transcript
```{r fig.height=1.75, fig.width=1.4}

#subset normalized counts table
regression_lcl_Xgenes <- regr_lcl_X
regression_fib_Xgenes <- regr_fib_X

regression_lcl_Xgenes$transcript <- row.names(regression_lcl_Xgenes)
regression_fib_Xgenes$transcript <- row.names(regression_fib_Xgenes)
merged_regression <- full_join(regression_lcl_Xgenes, regression_fib_Xgenes, by = "transcript")
row.names(merged_regression) <- merged_regression$transcript 
merged_regression$transcript <- NULL
#write.table(merged_regression, file = "merged_LCL_Fib_regresssion.txt", quote = FALSE, sep = "\t")


#LCL VERSION
sleuth_lcl_normcounts_X$transcript_id <- rownames(sleuth_lcl_normcounts_X) 
regression_lcl_Xgenes$transcript_id <- rownames(regression_lcl_Xgenes)
sleuth_lcl_normcounts_X_joined <- left_join(regression_lcl_Xgenes, sleuth_lcl_normcounts_X, by = "transcript_id")
pivot_sleuth_lcl_normcounts_X_joined <- pivot_longer(sleuth_lcl_normcounts_X_joined, (ncol(regression_lcl_Xgenes)+1):ncol(sleuth_lcl_normcounts_X_joined),  names_to = 'sample', values_to = 'counts')
merged_lcl_X_tograph <-merge(pivot_sleuth_lcl_normcounts_X_joined, metadata,  by = 'sample')

#FIB VERSION
sleuth_fib_normcounts_X$transcript_id <- rownames(sleuth_fib_normcounts_X) 
regression_fib_Xgenes$transcript_id <- rownames(regression_fib_Xgenes)
sleuth_fib_normcounts_X_joined <- left_join(regression_fib_Xgenes, sleuth_fib_normcounts_X, by = "transcript_id")
pivot_sleuth_fib_normcounts_X_joined <- pivot_longer(sleuth_fib_normcounts_X_joined, (ncol(regression_fib_Xgenes)+1):ncol(sleuth_fib_normcounts_X_joined),  names_to = 'sample', values_to = 'counts')
merged_lcl_X_tograph <-merge(pivot_sleuth_fib_normcounts_X_joined, metadata,  by = 'sample')


adjustLabel = "topRight"

gene <- "CA5B" #adjust for the gene whose transcript you want to plot!
ids <- c("ENST00000454127") #adjust for the transcript you want to plot!
myOrange <- "#d95f0290"
for (id in ids){
  merged_lcl_X_tograph1 <- merged_lcl_X_tograph %>% filter(transcript_id == id)
  merged_lcl_X_tograph1$x_count <- as.character(merged_lcl_X_tograph1$x_count)
  merged_lcl_X_tograph1$y_count <- as.character(merged_lcl_X_tograph1$y_count)
  merged_lcl_X_tograph1$x_count <- as.numeric(merged_lcl_X_tograph1$x_count) - 1
  merged_lcl_X_tograph1$counts <- merged_lcl_X_tograph1$counts /1000 #normalized
  # print(merged_lcl_X_tograph1) 
  # print(max(merged_lcl_X_tograph1$counts))

  myInt <- merged_lcl_X_tograph1$intercept[1] / 1000
  mySlope <- merged_lcl_X_tograph1$x_coeff[1] / 1000
  myDelta <- formatC(merged_lcl_X_tograph1$deltaEx[1], digits = 3)
  myP <- formatC(merged_lcl_X_tograph1$x_adj_pval[1], digits=2, format = "e")

  if(adjustLabel == "bottomLeft"){
    yloc <- 0
    xloc <- -0.3
    myV <- 0
    myH <- 0
  } 
  if(adjustLabel == "bottomRight"){
    yloc <- 0
    xloc <- 3.3
    myV <- 0
    myH <- 1
  }
  if(adjustLabel == "topLeft"){
    yloc <-  1.163132  + 0.05
    xloc <- -0.3
    myV <- 1
    myH <- 0
  }
  if(adjustLabel == "topRight"){
    yloc <- 1.163132 #max(merged_lcl_X_tograph1$counts)
    xloc <- 3.3
    myV <- 1
    myH <- 1
  }
  
  myWidth=1.4
  myHeight=1.75
  
  
pdf(file=paste0(id,"_", gene, "Fib_linearPlot.pdf"), width=myWidth, height=myHeight)
  p <- ggplot(data=merged_lcl_X_tograph1, mapping = aes(x=x_count, y=counts)) +
    geom_jitter(color="#00000080", position=position_jitter(0.2), size=1, stroke=0) +
    geom_abline(slope = mySlope, intercept = myInt, color=myOrange, lty=2) +
    annotate(
      geom = "text", 
      x = xloc, 
      y = yloc,
      label = paste0('dEX=', myDelta, '\nFDR=', myP),
      hjust = myH, 
      vjust = myV,
      size=2
      ) +
    labs( 
      x= "Number of Chr X",
      y= paste0("Normalized read\ncounts (x",1000,")"),
      title = paste0("*",id,"*"," (","NPX",")")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    
    ylim(-0.01,(1.163132 + 0.05)) + 
    xlim(-0.3,3.3)
  
  print(p)
  dev.off()
}
```


