---
title: "Modeling sex chromosome expression"
output: html_notebook
---

## Setup
```{r}
myPath <-  #ADD PATH TO GITHUB FOLDER
suppressPackageStartupMessages(library("limma"))
suppressPackageStartupMessages(library("rlist"))
suppressPackageStartupMessages(library("broom"))

#Set up sex chromosome gene lists, using gencode v84 as "Gene" list
geneAnno <- read.delim(file=paste0(myPath,"Annotations/geneAnno_proteinCoding_all_20210503.txt"), stringsAsFactors = FALSE)
colnames(geneAnno)[9] <- "Gene"
geneAnno_x <- geneAnno[geneAnno$chr == "chrX",]
geneAnno_y <- geneAnno[geneAnno$chr == "chrY",]
X_genes_all <-as.character(geneAnno_x$Gene)
Y_genes_all <-as.character(geneAnno_y$Gene)
sexChrom_genes <- c(X_genes_all,Y_genes_all)
PAR1_genes <- geneAnno_x[geneAnno_x$start < 2691188,]$Gene
PAR2_genes <- c("SPRY3","VAMP7","IL9R","WASH6P")
PAR_genes_all <- c(PAR1_genes,PAR2_genes)
NPX_genes <- X_genes_all[! X_genes_all %in% PAR_genes_all]
NPY_genes <- Y_genes_all[! Y_genes_all %in% PAR_genes_all]
autosome_genes <- geneAnno[! geneAnno$chr %in% c("chrX","chrY"),"Gene"]

#which cell type to analyze?
myCell <- "LCL"
# myCell <- "Fib"

#read in the metadata table
metadata <-read.delim(paste0(myPath,"Linear_regressions/metadata.txt"), stringsAsFactors = FALSE)
#subtract 1 from X count to model number of additional X chromosomes
metadata$x_count <- metadata$x_count - 1

#restrict to cell type and karyotype
## CHANGE FOR FIBROBLAST!
metadata_myCell <- metadata[metadata$karyotype %in% c("X","XX","XXX","XXXX","XXXXY","XXXY","XXY","XXYY","XY","XYY","XYYYY") & metadata$cell_type == myCell & metadata$Technical_replicate == "N",]
rownames(metadata_myCell) <- metadata_myCell$sample

#restrict to male samples only for NPY analysis
metadata_myCell_male <- metadata[metadata$karyotype %in% c("XXXXY","XXXY","XXY","XXYY","XY","XYY","XYYYY") & metadata$cell_type == myCell & metadata$Technical_replicate == "N",]
rownames(metadata_myCell_male) <- metadata_myCell_male$sample

#subtract 1 from Y count to model number of additional Y chromosomes
metadata_myCell_male$y_count <- metadata_myCell_male$y_count - 1
```

#Processing raw RNA-seq data
```{r, eval=FALSE}
### The following code is used to run this analysis from the raw data. Skip below to begin with provided processed data file. ###
## You must obtain access to the raw data through dbGaP (accession # phs002481). Download the fastq files, and run kallisto using the following command:
# kallisto quant -i KALLISTO_INDEX_FILE -t 16 --bias --plaintext -o  /kallisto_OUTPUT_FOLDER/sampleName/ file1fastq.gz file2fastq.gz 

#Bring in list of Kallisto output files
# myFileTable <- read.table(file =  "PATH TO LIST OF FILES", row.names = 1, stringsAsFactors = FALSE)
# myFiles <- myFileTable$V2
# names(myFiles) <- rownames(myFileTable)
#subset file list for the samples I will use in this analysis
# myFiles_myCell <- myFiles[metadata_myCell$sample]
# myFiles_myCell_male <- myFiles[metadata_myCell_male$sample]

#run tximport to bring in mapped samples
# suppressPackageStartupMessages(library("tximport"))

# annofile <-read.delim(file=paste0(myPath,"Annotations/annotation_with_ERCC.txt"),sep = " ")
# tx2gene <-data.frame("TXNAME" = annofile$transcript_id,"GENEID" = annofile$gene_name)
# 
# txi_myCell <- tximport(myFiles_myCell, type = "kallisto", tx2gene = tx2gene)
# txi_myCell_male <- tximport(myFiles_myCell_male, type = "kallisto", tx2gene = tx2gene)
# 
# #Adjust sample tables to match list of samples in txi
# sample_table_par_npx <- metadata_myCell[colnames(txi_myCell$counts), c("x_count","y_count","batch_libprep")]
# write.table(sample_table_par_npx, file="sample_table_par_npx.txt", quote=FALSE, row.names=TRUE, col.names = TRUE, sep="\t")
# sample_table_npy <- metadata_myCell_male[colnames(txi_myCell_male$counts), c("x_count","y_count","batch_libprep")]
# write.table(sample_table_npy, file="sample_table_npy.txt", quote=FALSE, row.names=TRUE, col.names = TRUE, sep="\t")

# #use scientific notation
# options(scipen = 0, digits = 5) 

# #For PAR and NPX:
# #make sure x_count and y_count are numeric, and batch_libprep is a factor
# class(sample_table_par_npx$x_count)
# class(sample_table_par_npx$y_count)
# sample_table_par_npx$batch_libprep <- as.factor(sample_table_par_npx$batch_libprep)
# class(sample_table_par_npx$batch_libprep)
#
# Get list of expressed genes in the cell type of interest
# XX_sample_names <- metadata_myCell[metadata_myCell$karyotype == "XX", "sample"]
# XY_sample_names <- metadata_myCell[metadata_myCell$karyotype == "XY", "sample"]
# tpm1_genes_xx_xy_median <- function(tximport_file, XX_samples, XY_samples) {
#   kallisto_output_TPM_XY <- tximport_file$abundance[,c(XY_samples)] 
#   kallisto_output_TPM_XX <- tximport_file$abundance[,c(XX_samples)] 
#   median_TPM_XY <- apply(kallisto_output_TPM_XY,1,median)
#   median_TPM_XX <- apply(kallisto_output_TPM_XX,1,median)
#   #Put all of the medians together in a dataframe of expression per karyotype - numeric
#   expression_perKaryotype <- data.frame(median_TPM_XY, median_TPM_XX)
#   #Isolate the rows of the expression_TPM dataset that have at least one karyotype passing the threshold
#   expressedGenes <- expression_perKaryotype[expression_perKaryotype$median_TPM_XY >= 1 | expression_perKaryotype$median_TPM_XX >= 1,]
#   #Restrict by gene annotation file
#   expressedGenes <- expressedGenes[rownames(expressedGenes) %in% geneAnno$Gene,]
#   #write.table(expressedGenes, file=paste0("expressedGenes_TPM_",tximport_file,".txt"),quote=FALSE,sep="\t")
#   #See how many genes are expressed in at least one karyotype
#   num_expressedGenes <- dim(expressedGenes)[1] 
#   print(paste("Number of 'expressed' genes with median TPM >= 1 in either XX or XY samples: ", as.character(num_expressedGenes), sep=""))
#    
#   #Find expressed sex chromosome genes
#   expSexChromGenes <- expressedGenes[rownames(expressedGenes) %in% sexChrom_genes,]
#   print(paste("Number of 'expressed' sex chromosome genes: ", as.character(dim(expSexChromGenes)[1]), sep=""))
#   #write.table(expSexChromGenes, file="expressedSexChromGenes_TPM.txt",quote=FALSE,sep="\t")
#   
#   #Find expressed Y chromosome genes
#   expYGenes <- expressedGenes[rownames(expressedGenes) %in% Y_genes_all,]
#   print(paste("Number of 'expressed' Y chromosome genes: ", as.character(dim(expYGenes)[1]), sep=""))
#   #write.table(expYGenes, file="expressedYgenes_TPM.txt",quote=FALSE,sep="\t")
#   
#   #Find expressed X chromosome genes
#   expXGenes <- expressedGenes[rownames(expressedGenes) %in% X_genes_all,]
#   print(paste("Number of 'expressed' X chromosome genes: ", as.character(dim(expXGenes)[1]), sep=""))
#   #write.table(expXGenes, file="expressedXgenes_TPM.txt",quote=FALSE,sep="\t")
#   
#   #Find expressed PAR genes
#   expPARGenes <- expressedGenes[rownames(expressedGenes) %in% PAR_genes_all,]
#   print(paste("Number of 'expressed' PAR genes (small set): ", as.character(dim(expPARGenes)[1]), sep=""))
#   #write.table(expPARGenes, file="expressedPARgenes_TPM.txt",quote=FALSE,sep="\t")
#   
#   #Find expressed autosome genes
#   expAutoGenes <- expressedGenes[rownames(expressedGenes) %in% autosome_genes,]
#   print(paste("Number of 'expressed' autosome genes: ", as.character(dim(expAutoGenes)[1]), sep=""))
#   #write.table(expAutoGenes, file="expressedAutosomeGenes_TPM.txt",quote=FALSE,sep="\t")
#   
#   myList <- list("expressedGenes" = rownames(expressedGenes), "expSexChromGenes" = rownames(expSexChromGenes), 
#                  "expYGenes" = rownames(expYGenes), "expXGenes" = rownames(expXGenes), "expPARGenes" = rownames(expPARGenes),
#                  "expAutoGenes" = rownames(expAutoGenes))
#   
#   return(myList)
# }
# 
# myCell_expressed <- tpm1_genes_xx_xy_median(tximport_file = txi_myCell,XX_samples = XX_sample_names, XY_samples = XY_sample_names )
# save(myCell_expressed, file="CELLTYPE_expressed.rda")
# 
# #run deseq to get the normalized counts matrix across all samples.
# suppressPackageStartupMessages(library("DESeq2"))
#
# dds_myCell_par_npx <- DESeqDataSetFromTximport(txi = txi_myCell, colData = sample_table_par_npx, 
#                                             design = ~ x_count + y_count + batch_libprep)
# dds_myCell_par_npx <- estimateSizeFactors(dds_myCell_par_npx)
# normCounts <- counts(dds_myCell_par_npx, normalized=TRUE)
# 
# #get normalized counts for expressed X genes
# normCounts_expXgenes <- normCounts[rownames(normCounts) %in% myCell_expressed$expXGenes,]
# write.table(x=normCounts_expXgenes, file="CELLTYPE_normCounts_expXgenes.txt", quote=FALSE, sep="\t", row.names=TRUE, col.names = TRUE)

# #get normalized counts for expressed Y genes
# sample_table_npy <- metadata_myCell_male[colnames(txi_myCell_male$counts), c("x_count","y_count","batch_libprep")]
# normCounts_expYgenes <- normCounts[rownames(normCounts) %in% myCell_expressed$expYGenes,colnames(normCounts) %in% rownames(sample_table_npy)]
# write.table(x=normCounts_expYgenes, file = "CELLTYPE_normCounts_expYgenes.txt", quote=FALSE, sep="\t", row.names=TRUE, col.names = TRUE)
```


#Run linear regressions for X chromosome genes
```{r}
## START HERE WITH PROCESSED NORMALIZED COUNTS DATA
#Example with LCLs
normCounts_expXgenes <- read.delim(file=paste0(myPath,"Linear_regressions/LCLs/lcl_normCounts_expXgenes.txt"), stringsAsFactors = FALSE, check.names = FALSE)

exp_NPX_genes <- intersect(rownames(normCounts_expXgenes),NPX_genes)
exp_PAR1_genes <- intersect(rownames(normCounts_expXgenes), PAR1_genes)
exp_PAR2_genes <- intersect(rownames(normCounts_expXgenes), PAR2_genes)

#Adjust sample tables to match list of samples in txi
sample_table_par_npx <- metadata_myCell[colnames(normCounts_expXgenes), c("x_count","y_count","batch_libprep")]

myResults_par_npx <- data.frame(intercept=numeric(),intErr = numeric(),xbeta=numeric(), xpval = numeric(), xErr = numeric(),ybeta=numeric(), ypval=numeric(),yErr=numeric(), adj_rsq=numeric(), pval=numeric())
region_res <- character()
for(gene in rownames(normCounts_expXgenes)){
  options(scipen = 0, digits = 5) 
    #pull out data
    myGeneData <- t(normCounts_expXgenes[gene,])
    all_Data <- cbind(sample_table_par_npx, myGeneData)
    colnames(all_Data) <- c("x_count", "y_count", "batch_libprep", "gene_expression")
    #do a linear regression
    myFormula <- formula(gene_expression ~ x_count + y_count + batch_libprep)
    myFormula_int <- formula(gene_expression ~ x_count + y_count )
    mylm <- lm(myFormula, data=all_Data)
    mylm_int <- lm(myFormula_int, data=all_Data)
    #pull out intercept, betas, p-values and add to table
    intercept <- summary(mylm_int)$coefficients["(Intercept)",1]
    interr <- summary(mylm_int)$coefficients["(Intercept)",2]
    xbeta <- summary(mylm)$coefficients["x_count",1]
    xpval <- summary(mylm)$coefficients["x_count",4]
    xerr <- summary(mylm)$coefficients["x_count",2]
    ybeta <- summary(mylm)$coefficients["y_count",1]
    ypval <- summary(mylm)$coefficients["y_count",4]
    yerr <- summary(mylm)$coefficients["y_count",2]
    adj_rsq <- glance(mylm)$adj.r.squared
    pval <- glance(mylm)$p.value
  
    #label genes with regions  
  if(gene %in% exp_NPX_genes){
      region <- "NPX"
    }
    if(gene %in% exp_PAR1_genes){
      region <- "PAR1"
    }
    if(gene %in% exp_PAR2_genes){
      region <- "PAR2"
    }
  #build results into table  
  myResults_par_npx[gene,] <- c(intercept,interr,xbeta, xpval, xerr,ybeta, ypval, yerr, adj_rsq,pval)
  region_res <- c(region_res, region)
}

myResults_par_npx <- cbind(myResults_par_npx,region_res)


#correct for multiple hypothesis testing and set signficance threshold to padj < 0.05
bh_adj <- p.adjust(myResults_par_npx$xpval, method="BH")
bh_adj_y <- p.adjust(myResults_par_npx$ypval, method="BH")
bh_adj_reg <- p.adjust(myResults_par_npx$pval, method="BH")
deltaEx <- myResults_par_npx$xbeta/myResults_par_npx$intercept
deltaEx_err <- sqrt(
  ((myResults_par_npx$xbeta ^ 2)/(myResults_par_npx$intercept ^ 2)) * 
  (
    ((myResults_par_npx$xErr ^ 2)/(myResults_par_npx$xbeta ^ 2)) + 
      ((myResults_par_npx$intErr ^ 2)/(myResults_par_npx$intercept ^ 2))
    )
  )
deltaEy <- myResults_par_npx$ybeta/myResults_par_npx$intercept
deltaEy_err <- sqrt(
  ((myResults_par_npx$ybeta ^ 2)/(myResults_par_npx$intercept ^ 2)) * 
  (
    ((myResults_par_npx$yErr ^ 2)/(myResults_par_npx$ybeta ^ 2)) + 
      ((myResults_par_npx$intErr ^ 2)/(myResults_par_npx$intercept ^ 2))
    )
  )

#put together final table
myResults_npx_par_deltaEx_y <- data.frame(row.names= rownames(myResults_par_npx),
                          "gene"= rownames(myResults_par_npx),
                          "intercept" = myResults_par_npx$intercept,
                          "intercept_err" = myResults_par_npx$intErr,
                          "x_coeff" = myResults_par_npx$xbeta,
                          "x_coeff_err" = myResults_par_npx$xErr,
                          "x_pval" = myResults_par_npx$xpval,
                          "x_adj_pval" = bh_adj,
                          "y_coeff" = myResults_par_npx$ybeta,
                          "y_coeff_err" = myResults_par_npx$yErr,
                          "y_pval" = myResults_par_npx$ypval,
                          "y_adj_pval" = bh_adj_y,
                          "adj.r.sq" = myResults_par_npx$adj_rsq,
                          "reg_pval" = myResults_par_npx$pval,
                          "reg_adj_pval" = bh_adj_reg,
                          "deltaEx" = deltaEx,
                          "deltaEx_err" = deltaEx_err,
                          "deltaEy" = deltaEy,
                          "deltaEy_err" = deltaEy_err,
                          "region" = myResults_par_npx$region_res
                          )

#adjust for XIST:
XIST_int <- myResults_par_npx["XIST","xbeta"] + myResults_par_npx["XIST","intercept"]
deltaEx_XIST <- myResults_par_npx["XIST","xbeta"]/XIST_int
deltaEy_XIST <- myResults_par_npx["XIST","ybeta"]/XIST_int

deltaEx_err_XIST <- sqrt(
  ((myResults_par_npx["XIST","xbeta"] ^ 2)/(XIST_int ^ 2)) * 
  (
    ((myResults_par_npx["XIST","xErr"] ^ 2)/(myResults_par_npx["XIST","xbeta"] ^ 2)) + 
      (((myResults_par_npx["XIST","intErr"] ^ 2) + (myResults_par_npx["XIST","xErr"] ^ 2))/(XIST_int ^ 2))
    )
  )

deltaEy_err_XIST <- sqrt(
  ((myResults_par_npx["XIST","ybeta"] ^ 2)/(XIST_int ^ 2)) * 
  (
    ((myResults_par_npx["XIST","yErr"] ^ 2)/(myResults_par_npx["XIST","ybeta"] ^ 2)) + 
      (((myResults_par_npx["XIST","intErr"] ^2) + (myResults_par_npx["XIST","yErr"] ^ 2 ))/(XIST_int ^ 2))
    )
  )


myResults_npx_par_deltaEx_y["XIST",c("deltaEx","deltaEx_err","deltaEy","deltaEy_err")] <- c(        "deltaEx" = deltaEx_XIST, "deltaEx_err" = deltaEx_err_XIST,"deltaEy" = deltaEy_XIST,"deltaEy_err" = deltaEy_err_XIST)

#merge results table with annotations
myResults_npx_par_deltaEx_y_anno <- merge(x=myResults_npx_par_deltaEx_y, y=geneAnno_x, by.x="gene",by.y="Gene")
# write.table(x = myResults_npx_par_deltaEx_y_anno, file = "lcl_regression_results_npx_par.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```

### Linear modeling of NPY genes
```{r}
## START HERE WITH PROCESSED NORMALIZED COUNTS DATA
#Example with LCLs
normCounts_expYgenes <- read.delim(file=paste0(myPath,"Linear_regressions/LCLs/lcl_normCounts_expYgenes.txt"), stringsAsFactors = FALSE, check.names = FALSE)

sample_table_npy <- metadata_myCell_male[colnames(normCounts_expYgenes), c("x_count","y_count","batch_libprep")]

myResults_npy <- data.frame(intercept=numeric(),intErr = numeric(),xbeta=numeric(), xpval = numeric(), xErr = numeric(),ybeta=numeric(), ypval=numeric(),yErr=numeric(), adj_rsq=numeric() , pval=numeric())
for(gene in rownames(normCounts_expYgenes)){
    options(scipen = 0, digits = 5) 
    #pull out data
    myGeneData <- t(normCounts_expYgenes[gene,])
    all_Data <- cbind(sample_table_npy, myGeneData)
    colnames(all_Data) <- c("x_count", "y_count", "batch_libprep", "gene_expression")
    #do a linear regression
    myFormula <- formula(gene_expression ~ x_count + y_count + batch_libprep)
    myFormula_int <- formula(gene_expression ~ x_count + y_count)
    mylm <- lm(myFormula, data=all_Data)
    mylm_int <- lm(myFormula_int, data=all_Data)
    #pull out intercepts, betas, p-values and add to table
    intercept <- summary(mylm_int)$coefficients["(Intercept)",1]
    interr <- summary(mylm_int)$coefficients["(Intercept)",2]
    xbeta <- summary(mylm)$coefficients["x_count",1]
    xpval <- summary(mylm)$coefficients["x_count",4]
    xerr <- summary(mylm)$coefficients["x_count",2]
    ybeta <- summary(mylm)$coefficients["y_count",1]
    ypval <- summary(mylm)$coefficients["y_count",4]
    yerr <- summary(mylm)$coefficients["y_count",2]
    adj_rsq <- glance(mylm)$adj.r.squared
    pval <- glance(mylm)$p.value
    
  myResults_npy[gene,] <- c(intercept,interr,xbeta, xpval, xerr,ybeta, ypval, yerr, adj_rsq , pval)
}


#correct for multiple hypothesis testing and set signficance threshold to padj < 0.05
options(scipen = 0) 
bh_adj <- p.adjust(myResults_npy$xpval, method="BH")
bh_adj_y <- p.adjust(myResults_npy$ypval, method="BH")
bh_adj_reg <- p.adjust(myResults_npy$pval, method="BH")
deltaEx <- myResults_npy$xbeta/myResults_npy$intercept
deltaEx_err <- sqrt(((myResults_npy$xbeta ^ 2)/(myResults_npy$intercept ^ 2)) * 
  (((myResults_npy$xErr ^ 2)/(myResults_npy$xbeta ^ 2)) + ((myResults_npy$intErr ^ 2)/(myResults_npy$intercept ^ 2))))
deltaEy <- myResults_npy$ybeta/myResults_npy$intercept
deltaEy_err <- sqrt(((myResults_npy$ybeta ^ 2)/(myResults_npy$intercept ^ 2)) * 
  (((myResults_npy$yErr ^ 2)/(myResults_npy$ybeta ^ 2)) + ((myResults_npy$intErr ^ 2)/(myResults_npy$intercept ^ 2))))

#put together final table
myResults_npy_deltaEx_y <- data.frame(row.names= rownames(myResults_npy),
                          "gene" = rownames(myResults_npy),
                          "intercept" = myResults_npy$intercept,
                          "intercept_err" = myResults_npy$intErr,
                          "x_coeff" = myResults_npy$xbeta,
                          "x_coeff_err" = myResults_npy$xErr,
                          "x_pval" = myResults_npy$xpval,
                          "x_adj_pval" = bh_adj,
                          "y_coeff" = myResults_npy$ybeta,
                          "y_coeff_err" = myResults_npy$yErr,
                          "y_pval" = myResults_npy$ypval,
                          "y_adj_pval" = bh_adj_y,
                          "adj.r.sq" = myResults_npy$adj_rsq,
                          "reg_pval" = myResults_npy$pval,
                          "reg_adj_pval" = bh_adj_reg,
                          "deltaEx" = deltaEx,
                          "deltaEx_err" = deltaEx_err,
                          "deltaEy" = deltaEy,
                          "deltaEy_err" = deltaEy_err
                          )

myResults_npy_deltaEx_y_anno <- merge(x=myResults_npy_deltaEx_y, y=geneAnno_y, by.x="gene",by.y="Gene")
# write.table(x = myResults_npy_deltaEx_y_anno, file = "lcl_regression_results_npy.txt", quote = FALSE, sep = "\t", row.names = FALSE)
```

```{r}
sessionInfo()
```



