---
title: "Modeling sex chromosome expression"
output: html_notebook
---

## Setup
```{r}
myPath <-  #ADD PATH TO GITHUB FOLDER
suppressPackageStartupMessages(library("limma"))
suppressPackageStartupMessages(library("rlist"))
suppressPackageStartupMessages(library("broom"))
suppressPackageStartupMessages(library("ggplot2"))


#Set up sex chromosome gene lists, using gencode v84 as "Gene" list
geneAnno <- read.delim(file=paste0(myPath,"Annotations/geneAnno_proteinCoding_all_20210503.txt"), stringsAsFactors = FALSE)
colnames(geneAnno)[9] <- "Gene"
geneAnno_x <- geneAnno[geneAnno$chr == "chrX",]
geneAnno_y <- geneAnno[geneAnno$chr == "chrY",]
X_genes_all <-as.character(geneAnno_x$Gene)
Y_genes_all <-as.character(geneAnno_y$Gene)
sexChrom_genes <- c(X_genes_all,Y_genes_all)
PAR1_genes <- geneAnno_x[geneAnno_x$start < 2691188,]$Gene
PAR2_genes <- c("SPRY3","VAMP7","IL9R","WASH6P")
PAR_genes_all <- c(PAR1_genes,PAR2_genes)
NPX_genes <- X_genes_all[! X_genes_all %in% PAR_genes_all]
NPY_genes <- Y_genes_all[! Y_genes_all %in% PAR_genes_all]
autosome_genes <- geneAnno[! geneAnno$chr %in% c("chrX","chrY"),"Gene"]

#which cell type to analyze?
myCell <- "Fib"

#read in the metadata table
metadata <-read.delim(paste0(myPath,"Linear_regressions/metadata.txt"), stringsAsFactors = FALSE)
#subtract 1 from X count to model number of additional X chromosomes
metadata$x_count <- metadata$x_count - 1

#restrict to cell type and karyotype
metadata_myCell <- metadata[metadata$karyotype %in% c("X","XX","XXX","XXXX","XXXXY","XXXY","XXY","XXYY","XY","XYY","XYYYY") & metadata$cell_type == myCell & metadata$Technical_replicate == "N",]
rownames(metadata_myCell) <- metadata_myCell$sample

#restrict to male samples only for NPY analysis
metadata_myCell_male <- metadata[metadata$karyotype %in% c("XXXXY","XXXY","XXY","XXYY","XY","XYY","XYYYY") & metadata$cell_type == myCell & metadata$Technical_replicate == "N",]
rownames(metadata_myCell_male) <- metadata_myCell_male$sample

#subtract 1 from Y count to model number of additional Y chromosomes
metadata_myCell_male$y_count <- metadata_myCell_male$y_count - 1
```

#Functions
```{r}
myOrange <- "#d95f02"
myOrange_light <- "#d95f0220"
myPurple <- "#7570B3"
myPurple_light <- "#7570B320"


ggGenePlot_regLine_xcount <- function(Gene,all_data, normBy, prefix, myWidth=1.525, myHeight=1.75){
  ## Uncomment to troubleshoot
  # Gene <- "XIST"
  # normCounts <- lcl_normCounts
  # metadata <- lcl_metadata
  # resTable <- lcl_res
  # prefix<-"LCL"
  # adjustLabel <-"bottomRight"
  # normBy <- 1000
  # region<-"NPX"
  
  all_data$gene_expression <- all_data$gene_expression / normBy
  all_data$x_count <- all_data$x_count + 1
  
  pdf(file=paste0("Figures/",prefix,"_",Gene,"_linearPlot.pdf"), width=myWidth, height=myHeight)
  p <- ggplot(data=all_data, mapping = aes(x=x_count, y=gene_expression)) +
    geom_jitter(color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(method = "lm", color=myOrange, size=0.25, fill=myOrange_light, fullrange=TRUE) +
    # annotate(
    #   geom = "text", 
    #   x = xloc, 
    #   y = yloc,
    #   label = paste0('dEX=', myDelta, '\nFDR=', myP),
    #   hjust = myH, 
    #   vjust = myV,
    #   size=2
    #   ) +
    labs( 
      x= "Number of Chr X",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(-0.01,max(all_data$gene_expression)+ 0.5) + xlim(0.7,4.3)
  print(p)
  dev.off()
}

ggGenePlot_regLine_xcount_min_max <- function(Gene,all_data, normBy, prefix, myWidth=1.525, myHeight=1.75, myMin = -0.01, myMax=10){
  ## Uncomment to troubleshoot
  # Gene <- "XIST"
  # normCounts <- lcl_normCounts
  # metadata <- lcl_metadata
  # resTable <- lcl_res
  # prefix<-"LCL"
  # adjustLabel <-"bottomRight"
  # normBy <- 1000
  # region<-"NPX"
  
  all_data$gene_expression <- all_data$gene_expression / normBy
  all_data$x_count <- all_data$x_count + 1
  
  pdf(file=paste0("Figures/",prefix,"_",Gene,"_linearPlot.pdf"), width=myWidth, height=myHeight)
  p <- ggplot(data=all_data, mapping = aes(x=x_count, y=gene_expression)) +
    geom_jitter(color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(method = "lm", color=myOrange, size=0.25, fill=myOrange_light, fullrange=TRUE) +
    # annotate(
    #   geom = "text", 
    #   x = xloc, 
    #   y = yloc,
    #   label = paste0('dEX=', myDelta, '\nFDR=', myP),
    #   hjust = myH, 
    #   vjust = myV,
    #   size=2
    #   ) +
    labs( 
      x= "Number of Chr X",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(myMin,myMax) + xlim(0.7,4.3)
  print(p)
  dev.off()
}

ggGenePlot_regLine_xcount_XIST <- function(Gene,all_data, normBy, prefix, myWidth=1.525, myHeight=1.75, myMin = -0.5, myMax = 10){
  ## Uncomment to troubleshoot
  # Gene <- "XIST"
  # normCounts <- lcl_normCounts
  # metadata <- lcl_metadata
  # resTable <- lcl_res
  # prefix<-"LCL"
  # adjustLabel <-"bottomRight"
  # normBy <- 1000
  # region<-"NPX"
  
  all_data$gene_expression <- all_data$gene_expression / normBy
  all_data$x_count <- all_data$x_count + 1
  all_data_xist <- all_data[all_data$x_count > 1,]
  
  pdf(file=paste0("Figures/",prefix,"_",Gene,"_linearPlot.pdf"), width=myWidth, height=myHeight)
  p <- ggplot() +
    geom_jitter(data=all_data, mapping = aes(x=x_count, y=gene_expression), color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(data=all_data_xist, mapping = aes(x=x_count, y=gene_expression), method = "lm", color=myOrange, size=0.25, fill=myOrange_light) +
    # annotate(
    #   geom = "text", 
    #   x = xloc, 
    #   y = yloc,
    #   label = paste0('dEX=', myDelta, '\nFDR=', myP),
    #   hjust = myH, 
    #   vjust = myV,
    #   size=2
    #   ) +
    labs( 
      x= "Number of Chr X",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(myMin,myMax) + xlim(0.7,4.3)
  print(p)
  dev.off()
}

ggGenePlot_regLine_ycount <- function(Gene,male_data,all_data, normBy, prefix, myWidth=1.525, myHeight=1.75){

  all_data$gene_expression <- all_data$gene_expression / normBy

  male_data$gene_expression <- male_data$gene_expression / normBy
    male_data$y_count <- male_data$y_count + 1


  pdf(file=paste0("Figures/",prefix,"_",Gene,"_linearPlot.pdf"), width=myWidth, height=myHeight)
  p <- ggplot() +
    geom_jitter(data=all_data, mapping = aes(x=y_count, y=gene_expression), color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(data=male_data, mapping = aes(x=y_count, y=gene_expression), method = "lm", color= myPurple, size=0.25, fill=myPurple_light) +
    labs( 
      x= "Number of Chr Y",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(-0.01,max(all_data$gene_expression)+ 0.5) + xlim(-0.3,4.3)
  print(p)
  dev.off()
}


ggGenePlot_regLine_ycount_max_min <- function(Gene,male_data,all_data, normBy, prefix, myWidth=1.525, myHeight=1.75, myMin = -0.01, myMax=10){

  all_data$gene_expression <- all_data$gene_expression / normBy

  male_data$gene_expression <- male_data$gene_expression / normBy
    male_data$y_count <- male_data$y_count + 1


  pdf(file=paste0("Figures/",prefix,"_",Gene,"_linearPlot.pdf"), width=myWidth, height=myHeight)
  p <- ggplot() +
    geom_jitter(data=all_data, mapping = aes(x=y_count, y=gene_expression), color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(data=male_data, mapping = aes(x=y_count, y=gene_expression), method = "lm", color= myPurple, size=0.25, fill=myPurple_light) +
    labs( 
      x= "Number of Chr Y",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(myMin,myMax) + xlim(-0.3,4.3)
  print(p)
  dev.off()
}

ggGenePlot_regLine_ycount_max_min_noZero <- function(Gene,male_data,all_data, normBy, prefix, myWidth=1.75, myHeight=1.75, myMin = -0.01, myMax=10){

  all_data$gene_expression <- all_data$gene_expression / normBy

  male_data$gene_expression <- male_data$gene_expression / normBy
    male_data$y_count <- male_data$y_count + 1


  pdf(file=paste0("Figures/",prefix,"_",Gene,"_linearPlot.pdf"), width=myWidth, height=myHeight)
  p <- ggplot() +
    geom_jitter(data=male_data, mapping = aes(x=y_count, y=gene_expression), color="#00000080", position=position_jitter(0.2), size=1.5, stroke=0) +
    geom_smooth(data=male_data, mapping = aes(x=y_count, y=gene_expression), method = "lm", color= myPurple, size=0.25, fill=myPurple_light) +
    labs( 
      x= "Number of Chr Y",
      y= paste0("Normalized read\ncounts (x",normBy,")"),
      title = paste0("*",Gene,"*")
      ) +
    theme_classic(base_size = 8) +
    theme(
      plot.title = ggtext::element_markdown(hjust = 0.5),
      axis.text = element_text(color = "black"), 
      axis.ticks = element_line(color = "black", size = 0.25), 
      axis.line = element_line(size = 0.25),
      # aspect.ratio=1
      ) +
    ylim(myMin,myMax) + xlim(0.7,4.3)
  print(p)
  dev.off()
}
```

#Processing raw RNA-seq data
```{r, eval=FALSE}
### The following code is used to run this analysis from the raw data. Skip below to begin with provided processed data file. ###
## You must obtain access to the raw data through dbGaP (accession # phs002481). Download the fastq files, and run kallisto using the following command:
# kallisto quant -i KALLISTO_INDEX_FILE -t 16 --bias --plaintext -o  /kallisto_OUTPUT_FOLDER/sampleName/ file1fastq.gz file2fastq.gz 

#Bring in list of Kallisto output files
myFileTable <- read.table(file =  "PATH TO LIST OF FILES", row.names = 1, stringsAsFactors = FALSE)
myFiles <- myFileTable$V2
names(myFiles) <- rownames(myFileTable)
#subset file list for the samples I will use in this analysis
myFiles_myCell <- myFiles[metadata_myCell$sample]
myFiles_myCell_male <- myFiles[metadata_myCell_male$sample]

#run tximport to bring in mapped samples
suppressPackageStartupMessages(library("tximport"))

annofile <-read.delim(file=paste0(myPath,"Annotations/annotation_with_ERCC.txt"),sep = " ")
tx2gene <-data.frame("TXNAME" = annofile$transcript_id,"GENEID" = annofile$gene_name)
# tx2gene_test <-data.frame("TXNAME" = annofile$transcript_id, "TXTYPE"=annofile$transcript_type,"GENEID" = annofile$gene_name, "GENETYPE"=annofile$gene_type)
# tx2gene_test_pc <- tx2gene_test[tx2gene_test$GENETYPE == "protein_coding",]
# tx2gene_test_pc_nonTx <- tx2gene_test_pc[! tx2gene_test_pc$TXTYPE == "protein_coding",]


txi_myCell <- tximport(myFiles_myCell, type = "kallisto", tx2gene = tx2gene)
txi_myCell_male <- tximport(myFiles_myCell_male, type = "kallisto", tx2gene = tx2gene)

#Adjust sample tables to match list of samples in txi
sample_table_par_npx <- metadata_myCell[colnames(txi_myCell$counts), c("x_count","y_count","batch_libprep")]
write.table(sample_table_par_npx, file="sample_table_par_npx.txt", quote=FALSE, row.names=TRUE, col.names = TRUE, sep="\t")
sample_table_npy <- metadata_myCell_male[colnames(txi_myCell_male$counts), c("x_count","y_count","batch_libprep")]
write.table(sample_table_npy, file="sample_table_npy.txt", quote=FALSE, row.names=TRUE, col.names = TRUE, sep="\t")

#use scientific notation
options(scipen = 0, digits = 5)

#For PAR and NPX:
#make sure x_count and y_count are numeric, and batch_libprep is a factor
class(sample_table_par_npx$x_count)
class(sample_table_par_npx$y_count)
sample_table_par_npx$batch_libprep <- as.factor(sample_table_par_npx$batch_libprep)
class(sample_table_par_npx$batch_libprep)

# #Get list of expressed genes in the cell type of interest:
XX_sample_names <- metadata_myCell[metadata_myCell$karyotype == "XX", "sample"]
XY_sample_names <- metadata_myCell[metadata_myCell$karyotype == "XY", "sample"]
tpm1_genes_xx_xy_median <- function(tximport_file, XX_samples, XY_samples) {
  kallisto_output_TPM_XY <- tximport_file$abundance[,c(XY_samples)]
  kallisto_output_TPM_XX <- tximport_file$abundance[,c(XX_samples)]
  median_TPM_XY <- apply(kallisto_output_TPM_XY,1,median)
  median_TPM_XX <- apply(kallisto_output_TPM_XX,1,median)
  #Put all of the medians together in a dataframe of expression per karyotype - numeric
  expression_perKaryotype <- data.frame(median_TPM_XY, median_TPM_XX)
  #Isolate the rows of the expression_TPM dataset that have at least one karyotype passing the threshold
  expressedGenes <- expression_perKaryotype[expression_perKaryotype$median_TPM_XY >= 1 | expression_perKaryotype$median_TPM_XX >= 1,]
  #Restrict by gene annotation file
  expressedGenes <- expressedGenes[rownames(expressedGenes) %in% geneAnno$Gene,]
  #write.table(expressedGenes, file=paste0("expressedGenes_TPM_",tximport_file,".txt"),quote=FALSE,sep="\t")
  #See how many genes are expressed in at least one karyotype
  num_expressedGenes <- dim(expressedGenes)[1]
  print(paste("Number of 'expressed' genes with median TPM >= 1 in either XX or XY samples: ", as.character(num_expressedGenes), sep=""))

  #Find expressed sex chromosome genes
  expSexChromGenes <- expressedGenes[rownames(expressedGenes) %in% sexChrom_genes,]
  print(paste("Number of 'expressed' sex chromosome genes: ", as.character(dim(expSexChromGenes)[1]), sep=""))
  #write.table(expSexChromGenes, file="expressedSexChromGenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed Y chromosome genes
  expYGenes <- expressedGenes[rownames(expressedGenes) %in% Y_genes_all,]
  print(paste("Number of 'expressed' Y chromosome genes: ", as.character(dim(expYGenes)[1]), sep=""))
  #write.table(expYGenes, file="expressedYgenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed X chromosome genes
  expXGenes <- expressedGenes[rownames(expressedGenes) %in% X_genes_all,]
  print(paste("Number of 'expressed' X chromosome genes: ", as.character(dim(expXGenes)[1]), sep=""))
  #write.table(expXGenes, file="expressedXgenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed PAR genes
  expPARGenes <- expressedGenes[rownames(expressedGenes) %in% PAR_genes_all,]
  print(paste("Number of 'expressed' PAR genes (small set): ", as.character(dim(expPARGenes)[1]), sep=""))
  #write.table(expPARGenes, file="expressedPARgenes_TPM.txt",quote=FALSE,sep="\t")

  #Find expressed autosome genes
  expAutoGenes <- expressedGenes[rownames(expressedGenes) %in% autosome_genes,]
  print(paste("Number of 'expressed' autosome genes: ", as.character(dim(expAutoGenes)[1]), sep=""))
  #write.table(expAutoGenes, file="expressedAutosomeGenes_TPM.txt",quote=FALSE,sep="\t")

  myList <- list("expressedGenes" = rownames(expressedGenes), "expSexChromGenes" = rownames(expSexChromGenes),
                 "expYGenes" = rownames(expYGenes), "expXGenes" = rownames(expXGenes), "expPARGenes" = rownames(expPARGenes),
                 "expAutoGenes" = rownames(expAutoGenes))

  return(myList)
}

fib_expressed <- tpm1_genes_xx_xy_median(tximport_file = txi_myCell,XX_samples = XX_sample_names, XY_samples = XY_sample_names )
save(fib_expressed, file=paste0(myPath,"Linear_regressions/Fibroblasts/fib_expressed.rda"))

# Bring in list of expressed genes in the cell type of interest
# load(file=paste0(myPath,"Linear_regressions/Fibroblasts/fib_expressed.rda"))
myCell_expressed <- fib_expressed


#run deseq to get the normalized counts matrix across all samples.
suppressPackageStartupMessages(library("DESeq2"))

dds_myCell_par_npx <- DESeqDataSetFromTximport(txi = txi_myCell, colData = sample_table_par_npx,
                                            design = ~ x_count + y_count + batch_libprep)
dds_myCell_par_npx <- estimateSizeFactors(dds_myCell_par_npx)
normCounts <- counts(dds_myCell_par_npx, normalized=TRUE)

#get normalized counts for expressed X genes
normCounts_expXgenes <- normCounts[rownames(normCounts) %in% myCell_expressed$expXGenes,]
write.table(x=normCounts_expXgenes, file=paste0(myCell,"_normCounts_expXgenes.txt"), quote=FALSE, sep="\t", row.names=TRUE, col.names = TRUE)


#get normalized counts for expressed Y genes
sample_table_npy_all <- metadata_myCell[colnames(txi_myCell$counts), c("x_count","y_count","batch_libprep")]
normCounts_expYgenes_all <- normCounts[rownames(normCounts) %in% myCell_expressed$expYGenes,colnames(normCounts) %in% rownames(sample_table_npy_all)]
write.table(x=normCounts_expYgenes_all, file = paste0(myCell,"_normCounts_expYgenes_all.txt"), quote=FALSE, sep="\t", row.names=TRUE, col.names = TRUE)

sample_table_npy_male <- metadata_myCell_male[colnames(txi_myCell_male$counts), c("x_count","y_count","batch_libprep")]
normCounts_expYgenes_male <- normCounts[rownames(normCounts) %in% myCell_expressed$expYGenes,colnames(normCounts) %in% rownames(sample_table_npy_male)]
write.table(x=normCounts_expYgenes_male, file = paste0(myCell,"_normCounts_expYgenes_male.txt"), quote=FALSE, sep="\t", row.names=TRUE, col.names = TRUE)
```


#Run linear regressions for X chromosome genes
```{r}
## START HERE WITH PROCESSED NORMALIZED COUNTS DATA
normCounts_expXgenes <- read.delim(file=paste0(myPath,"Linear_regressions/Fibroblasts/Fib_normCounts_expXgenes.txt"), stringsAsFactors = FALSE, check.names = FALSE)

exp_NPX_genes <- intersect(rownames(normCounts_expXgenes),NPX_genes)
exp_PAR1_genes <- intersect(rownames(normCounts_expXgenes), PAR1_genes)
exp_PAR2_genes <- intersect(rownames(normCounts_expXgenes), PAR2_genes)

#plot example genes
example_genes <- c("PRPS2","KDM5C","ZFX","F8")
lcl_top_15_up <- c("KDM5C","JPX","KDM6A","PNPLA4","PUDP",
                   "XIST","EIF1AX","AKAP17A","ZBED1","ZRSR2",
                   "SMC1A","GTPBP6","ZFX","RPS4X","SLC25A6")
lcl_not_sig_20 <- c("PHF8","ACSL4","NDUFA1","PRPS1","SLC25A53",
                    "TSC22D3","KIF4A","SLC10A3","EMD","IKBKG",
                    "UBL4A","FAM3A","TCEAL3","BCAP31","ARMCX3",
                    "CXorf40B","ENOX2","LAMP2","NXT2","RAB39B",
                    "RNF113A","RPL39","SH3KBP1","WDR45","MAGED2")
lcl_top_15_down <- c("CXorf57","TBC1D8B","F8","ZNF711","ARMCX5",
                     "USP11","HMGN5")

special <- c(example_genes,PAR1_genes,lcl_top_15_up,lcl_not_sig_20,lcl_top_15_down)

#Adjust sample tables to match list of samples in txi
sample_table_par_npx <- metadata_myCell[colnames(normCounts_expXgenes), c("x_count","y_count","batch_libprep")]

myResults_par_npx <- data.frame(intercept=numeric(),intErr = numeric(),xbeta=numeric(), xpval = numeric(), xErr = numeric(),ybeta=numeric(), ypval=numeric(),yErr=numeric(), adj_rsq=numeric(), pval=numeric())
region_res <- character()
for(gene in rownames(normCounts_expXgenes)){
options(scipen = 0, digits = 3) 
    #pull out data
    myGeneData <- t(normCounts_expXgenes[gene,])
    all_Data <- cbind(sample_table_par_npx, myGeneData)
    colnames(all_Data) <- c("x_count", "y_count", "batch_libprep", "gene_expression")
    #do a linear regression
    myFormula <- formula(gene_expression ~ x_count + y_count + batch_libprep)
    myFormula_int <- formula(gene_expression ~ x_count + y_count )
    mylm <- lm(myFormula, data=all_Data)
    mylm_int <- lm(myFormula_int, data=all_Data)
    #pull out intercept, betas, p-values and add to table
    intercept <- summary(mylm_int)$coefficients["(Intercept)",1]
    interr <- summary(mylm_int)$coefficients["(Intercept)",2]
    xbeta <- summary(mylm)$coefficients["x_count",1]
    xpval <- summary(mylm)$coefficients["x_count",4]
    xerr <- summary(mylm)$coefficients["x_count",2]
    ybeta <- summary(mylm)$coefficients["y_count",1]
    ypval <- summary(mylm)$coefficients["y_count",4]
    yerr <- summary(mylm)$coefficients["y_count",2]
    adj_rsq <- glance(mylm)$adj.r.squared
    pval <- glance(mylm)$p.value
  
    #label genes with regions  
  if(gene %in% exp_NPX_genes){
      region <- "NPX"
    }
    if(gene %in% exp_PAR1_genes){
      region <- "PAR1"
    }
    if(gene %in% exp_PAR2_genes){
      region <- "PAR2"
    }
  #build results into table  
  myResults_par_npx[gene,] <- c(intercept,interr,xbeta, xpval, xerr,ybeta, ypval, yerr, adj_rsq,pval)
  region_res <- c(region_res, region)
  
  #if gene in special list
  if(gene %in% special){
    print(gene)
      ggGenePlot_regLine_xcount(Gene = gene, all_data = all_Data,normBy = 1000, prefix = myCell)
  }

# #If you need to adjust plot limits for certain genes use the following:
  
#   if(gene=="SLC25A6"){
#     ggGenePlot_regLine_xcount_min_max(Gene = gene, all_data = all_Data, normBy = 1000, prefix = myCell, myMax = 50)
#   }
  
  
}


myResults_par_npx <- cbind(myResults_par_npx,region_res)


#correct for multiple hypothesis testing and set signficance threshold to padj < 0.05
bh_adj <- p.adjust(myResults_par_npx$xpval, method="BH")
bh_adj_y <- p.adjust(myResults_par_npx$ypval, method="BH")
bh_adj_reg <- p.adjust(myResults_par_npx$pval, method="BH")
deltaEx <- myResults_par_npx$xbeta/myResults_par_npx$intercept
deltaEx_err <- sqrt(
  ((myResults_par_npx$xbeta ^ 2)/(myResults_par_npx$intercept ^ 2)) * 
  (
    ((myResults_par_npx$xErr ^ 2)/(myResults_par_npx$xbeta ^ 2)) + 
      ((myResults_par_npx$intErr ^ 2)/(myResults_par_npx$intercept ^ 2))
    )
  )
deltaEx_95_CI_lo <- deltaEx - (1.96 * deltaEx_err)
deltaEx_95_CI_hi <- deltaEx + (1.96 * deltaEx_err)

deltaEy <- myResults_par_npx$ybeta/myResults_par_npx$intercept
deltaEy_err <- sqrt(
  ((myResults_par_npx$ybeta ^ 2)/(myResults_par_npx$intercept ^ 2)) * 
  (
    ((myResults_par_npx$yErr ^ 2)/(myResults_par_npx$ybeta ^ 2)) + 
      ((myResults_par_npx$intErr ^ 2)/(myResults_par_npx$intercept ^ 2))
    )
  )
deltaEy_95_CI_lo <- deltaEy - (1.96 * deltaEy_err)
deltaEy_95_CI_hi <- deltaEy + (1.96 * deltaEy_err)

#put together final table
myResults_npx_par_deltaEx_y <- data.frame(row.names= rownames(myResults_par_npx),
                          "gene"= rownames(myResults_par_npx),
                          "intercept" = myResults_par_npx$intercept,
                          "intercept_err" = myResults_par_npx$intErr,
                          "x_coeff" = myResults_par_npx$xbeta,
                          "x_coeff_err" = myResults_par_npx$xErr,
                          "x_pval" = myResults_par_npx$xpval,
                          "x_adj_pval" = bh_adj,
                          "y_coeff" = myResults_par_npx$ybeta,
                          "y_coeff_err" = myResults_par_npx$yErr,
                          "y_pval" = myResults_par_npx$ypval,
                          "y_adj_pval" = bh_adj_y,
                          "adj.r.sq" = myResults_par_npx$adj_rsq,
                          "reg_pval" = myResults_par_npx$pval,
                          "reg_adj_pval" = bh_adj_reg,
                          "deltaEx" = deltaEx,
                          "deltaEx_err" = deltaEx_err,
                          "deltaEx_95_CI_lo"= deltaEx_95_CI_lo,
                          "deltaEx_95_CI_hi"= deltaEx_95_CI_hi,
                          "deltaEy" = deltaEy,
                          "deltaEy_err" = deltaEy_err,
                          "deltaEy_95_CI_lo"= deltaEy_95_CI_lo,
                          "deltaEy_95_CI_hi"= deltaEy_95_CI_hi,
                          "region" = myResults_par_npx$region_res
                          )

#adjust for XIST:
XIST_int <- myResults_par_npx["XIST","xbeta"] + myResults_par_npx["XIST","intercept"]
deltaEx_XIST <- myResults_par_npx["XIST","xbeta"]/XIST_int
deltaEy_XIST <- myResults_par_npx["XIST","ybeta"]/XIST_int

deltaEx_err_XIST <- sqrt(
  ((myResults_par_npx["XIST","xbeta"] ^ 2)/(XIST_int ^ 2)) * 
  (
    ((myResults_par_npx["XIST","xErr"] ^ 2)/(myResults_par_npx["XIST","xbeta"] ^ 2)) + 
      (((myResults_par_npx["XIST","intErr"] ^ 2) + (myResults_par_npx["XIST","xErr"] ^ 2))/(XIST_int ^ 2))
    )
  )

deltaEx_95_CI_lo_xist <- deltaEx_XIST - (1.96 * deltaEx_err_XIST)
deltaEx_95_CI_hi_xist <- deltaEx_XIST + (1.96 * deltaEx_err_XIST)


deltaEy_err_XIST <- sqrt(
  ((myResults_par_npx["XIST","ybeta"] ^ 2)/(XIST_int ^ 2)) * 
  (
    ((myResults_par_npx["XIST","yErr"] ^ 2)/(myResults_par_npx["XIST","ybeta"] ^ 2)) + 
      (((myResults_par_npx["XIST","intErr"] ^2) + (myResults_par_npx["XIST","yErr"] ^ 2 ))/(XIST_int ^ 2))
    )
  )

deltaEy_95_CI_lo_xist <- deltaEy_XIST - (1.96 * deltaEy_err_XIST)
deltaEy_95_CI_hi_xist <- deltaEy_XIST + (1.96 * deltaEy_err_XIST)



myResults_npx_par_deltaEx_y["XIST",c("deltaEx","deltaEx_err","deltaEx_95_CI_lo","deltaEx_95_CI_hi","deltaEy","deltaEy_err",
                                     "deltaEy_95_CI_lo","deltaEy_95_CI_hi")] <- c("deltaEx" = deltaEx_XIST, 
                                                                                  "deltaEx_err" = deltaEx_err_XIST, 
                                                                                  "deltaEx_95_CI_lo" = deltaEx_95_CI_lo_xist,
                                                                                  "deltaEx_95_CI_hi" = deltaEx_95_CI_hi_xist,
                                                                                  "deltaEy" = deltaEy_XIST,
                                                                                  "deltaEy_err" = deltaEy_err_XIST,
                                                                                  "deltaEy_95_CI_lo" = deltaEy_95_CI_lo_xist,
                                                                                  "deltaEy_95_CI_hi" = deltaEy_95_CI_hi_xist)


write.table(x = myResults_npx_par_deltaEx_y, file = paste0(myCell,"_regression_results_npx_par.txt"), quote = FALSE, sep = "\t", row.names = FALSE)
```

## Linear regressions for NPY genes
```{r}
## START HERE WITH PROCESSED NORMALIZED COUNTS DATA
normCounts_expYgenes <- read.delim(file=paste0(myPath,"Linear_regressions/Fibroblasts/Fib_normCounts_expYgenes_male.txt"), stringsAsFactors = FALSE, check.names = FALSE)
normCounts_expYgenes_all <- read.delim(file=paste0(myPath,"Linear_regressions/Fibroblasts/Fib_normCounts_expYgenes_all.txt"), stringsAsFactors = FALSE, check.names = FALSE)

sample_table_npy <- metadata_myCell_male[colnames(normCounts_expYgenes), c("x_count","y_count","batch_libprep")]
sample_table_npy_all <- metadata_myCell[colnames(normCounts_expYgenes_all), c("x_count","y_count","batch_libprep")]

#plot example genes
special <- c(NPY_genes)

myResults_npy <- data.frame(intercept=numeric(),intErr = numeric(),xbeta=numeric(), xpval = numeric(), xErr = numeric(),ybeta=numeric(), ypval=numeric(),yErr=numeric(), adj_rsq=numeric() , pval=numeric())
for(gene in rownames(normCounts_expYgenes)){
    options(scipen = 0, digits = 5) 
    #pull out data
    myGeneData <- t(normCounts_expYgenes[gene,])
    male_Data <- cbind(sample_table_npy, myGeneData)
    colnames(male_Data) <- c("x_count", "y_count", "batch_libprep", "gene_expression")
    #do a linear regression
    myFormula <- formula(gene_expression ~ x_count + y_count + batch_libprep)
    myFormula_int <- formula(gene_expression ~ x_count + y_count)
    mylm <- lm(myFormula, data=male_Data)
    mylm_int <- lm(myFormula_int, data=male_Data)
    #pull out intercepts, betas, p-values and add to table
    intercept <- summary(mylm_int)$coefficients["(Intercept)",1]
    interr <- summary(mylm_int)$coefficients["(Intercept)",2]
    xbeta <- summary(mylm)$coefficients["x_count",1]
    xpval <- summary(mylm)$coefficients["x_count",4]
    xerr <- summary(mylm)$coefficients["x_count",2]
    ybeta <- summary(mylm)$coefficients["y_count",1]
    ypval <- summary(mylm)$coefficients["y_count",4]
    yerr <- summary(mylm)$coefficients["y_count",2]
    adj_rsq <- glance(mylm)$adj.r.squared
    pval <- glance(mylm)$p.value
    
  myResults_npy[gene,] <- c(intercept,interr,xbeta, xpval, xerr,ybeta, ypval, yerr, adj_rsq , pval)
  
  #if gene in special list
  if(gene %in% special){
    print(gene)
    #also bring in Y gene data for all samples
     myGeneData_all <- t(normCounts_expYgenes_all[gene,])
    all_Data <- cbind(sample_table_npy_all, myGeneData_all)
    colnames(all_Data) <- c("x_count", "y_count", "batch_libprep", "gene_expression")
      ggGenePlot_regLine_ycount(Gene = gene, male_data = male_Data, all_data = all_Data, normBy = 1000, prefix = myCell)
  }
  
    #If you need to adjust the axis limits, use this:
  # if(gene == "DDX3Y"){
  #   print(gene)
  #   myGeneData_all <- t(normCounts_expYgenes_all[gene,])
  #   all_Data <- cbind(sample_table_npy_all, myGeneData_all)
  #   colnames(all_Data) <- c("x_count", "y_count", "batch_libprep", "gene_expression")
  #   ggGenePlot_regLine_ycount_max_min(Gene = gene, male_data = male_Data, all_data = all_Data, normBy = 1000, prefix = myCell, myMax = 40)
  # }
  
}


#correct for multiple hypothesis testing and set signficance threshold to padj < 0.05
options(scipen = 0) 
bh_adj <- p.adjust(myResults_npy$xpval, method="BH")
bh_adj_y <- p.adjust(myResults_npy$ypval, method="BH")
bh_adj_reg <- p.adjust(myResults_npy$pval, method="BH")
deltaEx <- myResults_npy$xbeta/myResults_npy$intercept
deltaEx_err <- sqrt(((myResults_npy$xbeta ^ 2)/(myResults_npy$intercept ^ 2)) * 
  (((myResults_npy$xErr ^ 2)/(myResults_npy$xbeta ^ 2)) + ((myResults_npy$intErr ^ 2)/(myResults_npy$intercept ^ 2))))
deltaEx_95_CI_lo <- deltaEx - (1.96 * deltaEx_err)
deltaEx_95_CI_hi <- deltaEx + (1.96 * deltaEx_err)

deltaEy <- myResults_npy$ybeta/myResults_npy$intercept
deltaEy_err <- sqrt(((myResults_npy$ybeta ^ 2)/(myResults_npy$intercept ^ 2)) * 
  (((myResults_npy$yErr ^ 2)/(myResults_npy$ybeta ^ 2)) + ((myResults_npy$intErr ^ 2)/(myResults_npy$intercept ^ 2))))
deltaEy_95_CI_lo <- deltaEy - (1.96 * deltaEy_err)
deltaEy_95_CI_hi <- deltaEy + (1.96 * deltaEy_err)

#put together final table
myResults_npy_deltaEx_y <- data.frame(row.names= rownames(myResults_npy),
                          "gene" = rownames(myResults_npy),
                          "intercept" = myResults_npy$intercept,
                          "intercept_err" = myResults_npy$intErr,
                          "x_coeff" = myResults_npy$xbeta,
                          "x_coeff_err" = myResults_npy$xErr,
                          "x_pval" = myResults_npy$xpval,
                          "x_adj_pval" = bh_adj,
                          "y_coeff" = myResults_npy$ybeta,
                          "y_coeff_err" = myResults_npy$yErr,
                          "y_pval" = myResults_npy$ypval,
                          "y_adj_pval" = bh_adj_y,
                          "adj.r.sq" = myResults_npy$adj_rsq,
                          "reg_pval" = myResults_npy$pval,
                          "reg_adj_pval" = bh_adj_reg,
                          "deltaEx" = deltaEx,
                          "deltaEx_err" = deltaEx_err,
                          "deltaEx_95_CI_lo"= deltaEx_95_CI_lo,
                          "deltaEx_95_CI_hi"= deltaEx_95_CI_hi,
                          "deltaEy" = deltaEy,
                          "deltaEy_err" = deltaEy_err,
                          "deltaEy_95_CI_lo"= deltaEy_95_CI_lo,
                          "deltaEy_95_CI_hi"= deltaEy_95_CI_hi
                          )

write.table(x = myResults_npy_deltaEx_y, file = paste0(myCell,"_regression_results_npy.txt"), quote = FALSE, sep = "\t", row.names = FALSE)
```

```{r}
sessionInfo()
```



